

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>isambard.ampal.protein &mdash; ISAMBARD 1.4.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../search.html"/>
    <link rel="top" title="ISAMBARD 1.4.1 documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> ISAMBARD
          

          
          </a>

          
            
            
              <div class="version">
                2016.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial/tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api_reference.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../developer_guide.html">Developer Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../citing_isambard.html">Citing ISAMBARD</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">ISAMBARD</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>isambard.ampal.protein</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for isambard.ampal.protein</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;AMPAL objects that represent protein.&quot;&quot;&quot;</span>

<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">OrderedDict</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">numpy</span>

<span class="kn">from</span> <span class="nn">ampal.base_ampal</span> <span class="k">import</span> <span class="n">Polymer</span><span class="p">,</span> <span class="n">Monomer</span><span class="p">,</span> <span class="n">Atom</span>
<span class="kn">from</span> <span class="nn">ampal.ligands</span> <span class="k">import</span> <span class="n">Ligand</span><span class="p">,</span> <span class="n">LigandGroup</span>
<span class="kn">from</span> <span class="nn">ampal.assembly</span> <span class="k">import</span> <span class="n">Assembly</span>
<span class="kn">from</span> <span class="nn">ampal.pseudo_atoms</span> <span class="k">import</span> <span class="n">Primitive</span>
<span class="kn">from</span> <span class="nn">ampal.analyse_protein</span> <span class="k">import</span> <span class="p">(</span>
    <span class="n">make_primitive_extrapolate_ends</span><span class="p">,</span> <span class="n">measure_torsion_angles</span><span class="p">,</span> <span class="n">residues_per_turn</span><span class="p">,</span>
    <span class="n">polymer_to_reference_axis_distances</span><span class="p">,</span> <span class="n">crick_angles</span><span class="p">,</span> <span class="n">alpha_angles</span><span class="p">,</span>
    <span class="n">sequence_molecular_weight</span><span class="p">,</span> <span class="n">sequence_molar_extinction_280</span><span class="p">,</span>
    <span class="n">sequence_isoelectric_point</span><span class="p">,</span> <span class="n">measure_sidechain_torsion_angles</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">ampal.interactions</span> <span class="k">import</span> <span class="p">(</span>
    <span class="n">generate_covalent_bond_graph</span><span class="p">,</span> <span class="n">generate_bond_subgraphs_from_break</span><span class="p">,</span>
    <span class="n">find_covalent_bonds</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">external_programs.dssp</span> <span class="k">import</span> <span class="p">(</span>
    <span class="n">extract_all_ss_dssp</span><span class="p">,</span> <span class="n">run_dssp</span><span class="p">,</span> <span class="n">extract_solvent_accessibility_dssp</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">external_programs.naccess</span> <span class="k">import</span> <span class="n">run_naccess</span><span class="p">,</span> <span class="n">extract_residue_accessibility</span>
<span class="kn">from</span> <span class="nn">external_programs.scwrl</span> <span class="k">import</span> <span class="n">pack_sidechains</span>
<span class="kn">from</span> <span class="nn">settings</span> <span class="k">import</span> <span class="n">global_settings</span>
<span class="kn">from</span> <span class="nn">tools.amino_acids</span> <span class="k">import</span> <span class="p">(</span>
    <span class="n">get_aa_code</span><span class="p">,</span> <span class="n">get_aa_letter</span><span class="p">,</span> <span class="n">ideal_backbone_bond_lengths</span><span class="p">,</span>
    <span class="n">ideal_backbone_bond_angles</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">tools.geometry</span> <span class="k">import</span> <span class="p">(</span>
    <span class="n">Quaternion</span><span class="p">,</span> <span class="n">unit_vector</span><span class="p">,</span> <span class="n">dihedral</span><span class="p">,</span> <span class="n">find_transformations</span><span class="p">,</span> <span class="n">distance</span><span class="p">,</span>
    <span class="n">angle_between_vectors</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">tools.isambard_warnings</span> <span class="k">import</span> <span class="n">MalformedPDBWarning</span>


<div class="viewcode-block" id="find_ss_regions_polymer"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.find_ss_regions_polymer">[docs]</a><span class="k">def</span> <span class="nf">find_ss_regions_polymer</span><span class="p">(</span><span class="n">polymer</span><span class="p">,</span> <span class="n">ss</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns an `Assembly` of regions tagged as secondary structure.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    polymer : Polypeptide</span>
<span class="sd">        `Polymer` object to be searched secondary structure regions.</span>
<span class="sd">    ss : list</span>
<span class="sd">        List of secondary structure tags to be separate i.e. [&#39;H&#39;]</span>
<span class="sd">        would return helices, [&#39;H&#39;, &#39;E&#39;] would return helices</span>
<span class="sd">        and strands.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    fragments : Assembly</span>
<span class="sd">        `Assembly` containing a `Polymer` for each region of specified</span>
<span class="sd">        secondary structure.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ss</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">ss</span> <span class="o">=</span> <span class="p">[</span><span class="n">ss</span><span class="p">[:]]</span>
    <span class="n">tag_key</span> <span class="o">=</span> <span class="s1">&#39;secondary_structure&#39;</span>
    <span class="n">monomers</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">polymer</span> <span class="k">if</span> <span class="n">tag_key</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">tags</span><span class="o">.</span><span class="n">keys</span><span class="p">()]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">monomers</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">Assembly</span><span class="p">()</span>
    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ss</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="nb">all</span><span class="p">([</span><span class="n">m</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="n">tag_key</span><span class="p">]</span> <span class="o">==</span> <span class="n">ss</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">monomers</span><span class="p">])):</span>
        <span class="k">return</span> <span class="n">Assembly</span><span class="p">(</span><span class="n">polymer</span><span class="p">)</span>
    <span class="n">previous_monomer</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">fragment</span> <span class="o">=</span> <span class="n">Polypeptide</span><span class="p">(</span><span class="n">ampal_parent</span><span class="o">=</span><span class="n">polymer</span><span class="p">)</span>
    <span class="n">fragments</span> <span class="o">=</span> <span class="n">Assembly</span><span class="p">()</span>
    <span class="n">poly_id</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">monomer</span> <span class="ow">in</span> <span class="n">monomers</span><span class="p">:</span>
        <span class="n">current_monomer</span> <span class="o">=</span> <span class="n">monomer</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="n">tag_key</span><span class="p">]</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">current_monomer</span> <span class="o">==</span> <span class="n">previous_monomer</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="ow">not</span> <span class="n">previous_monomer</span><span class="p">):</span>
            <span class="n">fragment</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">monomer</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">previous_monomer</span> <span class="ow">in</span> <span class="n">ss</span><span class="p">:</span>
                <span class="n">fragment</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="n">tag_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">monomer</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="n">tag_key</span><span class="p">]</span>
                <span class="n">fragment</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="n">poly_id</span> <span class="o">+</span> <span class="mi">65</span><span class="p">)</span>
                <span class="n">fragments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fragment</span><span class="p">)</span>
                <span class="n">poly_id</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">fragment</span> <span class="o">=</span> <span class="n">Polypeptide</span><span class="p">(</span><span class="n">ampal_parent</span><span class="o">=</span><span class="n">polymer</span><span class="p">)</span>
            <span class="n">fragment</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">monomer</span><span class="p">)</span>
        <span class="n">previous_monomer</span> <span class="o">=</span> <span class="n">monomer</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="n">tag_key</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">fragments</span></div>


<div class="viewcode-block" id="flat_list_to_polymer"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.flat_list_to_polymer">[docs]</a><span class="k">def</span> <span class="nf">flat_list_to_polymer</span><span class="p">(</span><span class="n">atom_list</span><span class="p">,</span> <span class="n">atom_group_s</span><span class="o">=</span><span class="mi">4</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Takes a flat list of atomic coordinates and converts it to a `Polymer`.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    atom_list : [Atom]</span>
<span class="sd">        Flat list of coordinates.</span>
<span class="sd">    atom_group_s : int, optional</span>
<span class="sd">        Size of atom groups.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    polymer : Polypeptide</span>
<span class="sd">        `Polymer` object containing atom coords converted `Monomers`.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        Raised if `atom_group_s` != 4 or 5</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">atom_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;O&#39;</span><span class="p">,</span> <span class="s1">&#39;CB&#39;</span><span class="p">]</span>
    <span class="n">atom_elements</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;O&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">]</span>
    <span class="n">atoms_coords</span> <span class="o">=</span> <span class="p">[</span><span class="n">atom_list</span><span class="p">[</span><span class="n">x</span><span class="p">:</span><span class="n">x</span> <span class="o">+</span> <span class="n">atom_group_s</span><span class="p">]</span>
                    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">atom_list</span><span class="p">),</span> <span class="n">atom_group_s</span><span class="p">)]</span>
    <span class="n">atoms</span> <span class="o">=</span> <span class="p">[[</span><span class="n">Atom</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">atom_elements</span><span class="p">)]</span>
             <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">atoms_coords</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">atom_group_s</span> <span class="o">==</span> <span class="mi">5</span><span class="p">:</span>
        <span class="n">monomers</span> <span class="o">=</span> <span class="p">[</span><span class="n">Residue</span><span class="p">(</span><span class="n">OrderedDict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">atom_labels</span><span class="p">,</span> <span class="n">x</span><span class="p">)),</span> <span class="s1">&#39;ALA&#39;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">atoms</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">atom_group_s</span> <span class="o">==</span> <span class="mi">4</span><span class="p">:</span>
        <span class="n">monomers</span> <span class="o">=</span> <span class="p">[</span><span class="n">Residue</span><span class="p">(</span><span class="n">OrderedDict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">atom_labels</span><span class="p">,</span> <span class="n">x</span><span class="p">)),</span> <span class="s1">&#39;GLY&#39;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">atoms</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s1">&#39;Parameter atom_group_s must be 4 or 5 so atoms can be labeled correctly.&#39;</span><span class="p">)</span>
    <span class="n">polymer</span> <span class="o">=</span> <span class="n">Polypeptide</span><span class="p">(</span><span class="n">monomers</span><span class="o">=</span><span class="n">monomers</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">polymer</span></div>


<div class="viewcode-block" id="flat_list_to_dummy_chain"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.flat_list_to_dummy_chain">[docs]</a><span class="k">def</span> <span class="nf">flat_list_to_dummy_chain</span><span class="p">(</span><span class="n">atom_list</span><span class="p">,</span> <span class="n">atom_group_s</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Converts flat list of coordinates into dummy C-alpha carbons</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    atom_list : [Atom]</span>
<span class="sd">        Flat list of co-ordinates.</span>
<span class="sd">    atom_group_s : int, optional</span>
<span class="sd">        Size of atom groups.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    polymer : Polypeptide</span>
<span class="sd">        `Polymer` object containing atom coord converted `Monomers`</span>
<span class="sd">        with &#39;DUM&#39; atom name.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">atom_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span>
    <span class="n">atom_elements</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span>
    <span class="n">atoms_coords</span> <span class="o">=</span> <span class="p">[</span><span class="n">atom_list</span><span class="p">[</span><span class="n">x</span><span class="p">:</span><span class="n">x</span> <span class="o">+</span> <span class="n">atom_group_s</span><span class="p">]</span>
                    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">atom_list</span><span class="p">),</span> <span class="n">atom_group_s</span><span class="p">)]</span>
    <span class="n">atoms</span> <span class="o">=</span> <span class="p">[[</span><span class="n">Atom</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">atom_elements</span><span class="p">)]</span>
             <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">atoms_coords</span><span class="p">]</span>
    <span class="n">monomers</span> <span class="o">=</span> <span class="p">[</span><span class="n">Residue</span><span class="p">(</span><span class="n">OrderedDict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">atom_labels</span><span class="p">,</span> <span class="n">x</span><span class="p">)),</span> <span class="s1">&#39;DUM&#39;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">atoms</span><span class="p">]</span>
    <span class="n">polymer</span> <span class="o">=</span> <span class="n">Polypeptide</span><span class="p">(</span><span class="n">monomers</span><span class="o">=</span><span class="n">monomers</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">polymer</span></div>


<div class="viewcode-block" id="align"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.align">[docs]</a><span class="k">def</span> <span class="nf">align</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">mobile</span><span class="p">,</span> <span class="n">target_i</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">mobile_i</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Aligns one Polypeptide (mobile) to another (target).</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    This function directly modifies atoms of the mobile Polypeptide!</span>
<span class="sd">    It does not return a new object.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    target : Polypeptide</span>
<span class="sd">        Polypeptide to be aligned to.</span>
<span class="sd">    mobile : Polypeptide</span>
<span class="sd">        Polypeptide to be moved during alignment.</span>
<span class="sd">    target_i : int, optional</span>
<span class="sd">        Index of `Residue` in target to align to.</span>
<span class="sd">    mobile_i : int, optional</span>
<span class="sd">        Index of `Residue` in mobile to be aligned.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># First, align N-&gt;CA vectors.</span>
    <span class="n">s1</span><span class="p">,</span> <span class="n">e1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span> <span class="n">e2</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">_vector</span>
                      <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="n">mobile</span><span class="p">[</span><span class="n">mobile_i</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">],</span> <span class="n">mobile</span><span class="p">[</span><span class="n">mobile_i</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span>
                                <span class="n">target</span><span class="p">[</span><span class="n">target_i</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">],</span> <span class="n">target</span><span class="p">[</span><span class="n">target_i</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">]]]</span>
    <span class="n">translation</span><span class="p">,</span> <span class="n">angle</span><span class="p">,</span> <span class="n">axis</span><span class="p">,</span> <span class="n">point</span> <span class="o">=</span> <span class="n">find_transformations</span><span class="p">(</span>
        <span class="n">s1</span><span class="p">,</span> <span class="n">e1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span> <span class="n">e2</span><span class="p">,</span> <span class="n">radians</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="c1"># Rotation first, Then translation.</span>
    <span class="n">mobile</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="n">angle</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">,</span> <span class="n">point</span><span class="o">=</span><span class="n">point</span><span class="p">,</span> <span class="n">radians</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">mobile</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">vector</span><span class="o">=</span><span class="n">translation</span><span class="p">)</span>
    <span class="c1"># Second, rotate about N-&gt;CA axis to align CA-&gt;C vectors.</span>
    <span class="n">angle</span> <span class="o">=</span> <span class="n">dihedral</span><span class="p">(</span><span class="n">mobile</span><span class="p">[</span><span class="n">mobile_i</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">mobile</span><span class="p">[</span><span class="n">mobile_i</span><span class="p">]</span>
                     <span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">],</span> <span class="n">mobile</span><span class="p">[</span><span class="n">mobile_i</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span> <span class="n">target</span><span class="p">[</span><span class="n">target_i</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">])</span>
    <span class="n">axis</span> <span class="o">=</span> <span class="n">target</span><span class="p">[</span><span class="n">target_i</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">target</span><span class="p">[</span><span class="n">target_i</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">]</span>
    <span class="n">point</span> <span class="o">=</span> <span class="n">target</span><span class="p">[</span><span class="n">target_i</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span>
    <span class="n">mobile</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="n">angle</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">,</span> <span class="n">point</span><span class="o">=</span><span class="n">point</span><span class="p">)</span>
    <span class="k">return</span></div>


<div class="viewcode-block" id="Polypeptide"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide">[docs]</a><span class="k">class</span> <span class="nc">Polypeptide</span><span class="p">(</span><span class="n">Polymer</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Container for `Residues`, inherits from `Polymer`.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    monomers : Residue or [Residue], optional</span>
<span class="sd">        `Residue` or list containing `Residue` objects to form the</span>
<span class="sd">        `Polypeptide`.</span>
<span class="sd">    polymer_id : str, optional</span>
<span class="sd">        An ID that the user can use to identify the `Polypeptide`. This is</span>
<span class="sd">        used when generating a pdb file using `Polypeptide().pdb`.</span>
<span class="sd">    ampal_parent : ampal.Assembly, optional</span>
<span class="sd">        Reference to `Assembly` containing the `Polymer`.</span>
<span class="sd">    sl : int, optional</span>
<span class="sd">        The default smoothing level used when calculating the</span>
<span class="sd">        backbone primitive.</span>

<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    id : str</span>
<span class="sd">        `Polypeptide` ID</span>
<span class="sd">    ampal_parent : ampal.Assembly or None</span>
<span class="sd">        Reference to `Assembly` containing the `Polypeptide`</span>
<span class="sd">    molecule_type : str</span>
<span class="sd">        A description of the type of `Polymer` i.e. Protein, DNA etc.</span>
<span class="sd">    ligands : ampal.LigandGroup</span>
<span class="sd">        A `LigandGroup` containing all the `Ligands` associated with this</span>
<span class="sd">        `Polypeptide` chain.</span>
<span class="sd">    tags : dict</span>
<span class="sd">        A dictionary containing information about this AMPAL object.</span>
<span class="sd">        The tags dictionary is used by AMPAL to cache information</span>
<span class="sd">        about this object, but is also intended to be used by users</span>
<span class="sd">        to store any relevant information they have.</span>
<span class="sd">    sl : int</span>
<span class="sd">        The default smoothing level used when calculating the</span>
<span class="sd">        backbone primitive.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    TypeError</span>
<span class="sd">        `Polymer` type objects can only be initialised empty or using</span>
<span class="sd">        a `Monomer`.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">monomers</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">polymer_id</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">ampal_parent</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sl</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
            <span class="n">monomers</span><span class="o">=</span><span class="n">monomers</span><span class="p">,</span> <span class="n">polymer_id</span><span class="o">=</span><span class="n">polymer_id</span><span class="p">,</span> <span class="n">molecule_type</span><span class="o">=</span><span class="s1">&#39;protein&#39;</span><span class="p">,</span>
            <span class="n">ampal_parent</span><span class="o">=</span><span class="n">ampal_parent</span><span class="p">,</span> <span class="n">sl</span><span class="o">=</span><span class="n">sl</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__add__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">Polymer</span><span class="p">):</span>
            <span class="n">merged_polymer</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span> <span class="o">+</span> <span class="n">other</span><span class="o">.</span><span class="n">_monomers</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
                <span class="s1">&#39;Only Polymer objects may be merged with a Polymer.&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">Polypeptide</span><span class="p">(</span><span class="n">monomers</span><span class="o">=</span><span class="n">merged_polymer</span><span class="p">,</span> <span class="n">polymer_id</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">item</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">item</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">id_dict</span> <span class="o">=</span> <span class="p">{</span><span class="nb">str</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">id</span><span class="p">):</span> <span class="n">m</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">}</span>
            <span class="k">return</span> <span class="n">id_dict</span><span class="p">[</span><span class="n">item</span><span class="p">]</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">item</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="n">item</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">Polypeptide</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="n">item</span><span class="p">],</span> <span class="n">polymer_id</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">15</span><span class="p">:</span>
            <span class="n">seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">[:</span><span class="mi">12</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;...&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span>
        <span class="k">return</span> <span class="s1">&#39;&lt;Polypeptide containing </span><span class="si">{}</span><span class="s1"> </span><span class="si">{}</span><span class="s1">. Sequence: </span><span class="si">{}</span><span class="s1">&gt;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">),</span>
            <span class="s1">&#39;Residue&#39;</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="k">else</span> <span class="s1">&#39;Residues&#39;</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span>

<div class="viewcode-block" id="Polypeptide.get_slice_from_res_id"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.get_slice_from_res_id">[docs]</a>    <span class="k">def</span> <span class="nf">get_slice_from_res_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns a new `Polypeptide` containing the `Residues` in start/end range.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        start : str</span>
<span class="sd">            string representing start residue id (PDB numbering)</span>
<span class="sd">        end : str</span>
<span class="sd">            string representing end residue id (PDB numbering)</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        slice_polymer : Polymer</span>
<span class="sd">            Polymer containing the residue range specified by start-end</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">id_dict</span> <span class="o">=</span> <span class="p">{</span><span class="nb">str</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">id</span><span class="p">):</span> <span class="n">m</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">}</span>
        <span class="n">slice_polymer</span> <span class="o">=</span> <span class="n">Polypeptide</span><span class="p">(</span>
            <span class="p">[</span><span class="n">id_dict</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">end</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)],</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">slice_polymer</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">backbone</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns a new `Polymer` containing only the backbone atoms.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Metadata is not currently preserved from the parent object.</span>
<span class="sd">        Sequence data is retained, but only the main chain atoms are retained.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        bb_poly : Polypeptide</span>
<span class="sd">            Polymer containing only the backbone atoms of the original</span>
<span class="sd">            Polymer.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">bb_poly</span> <span class="o">=</span> <span class="n">Polypeptide</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">backbone</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">bb_poly</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">primitive</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Primitive of the backbone.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        This is the average of the positions of all the CAs in frames</span>
<span class="sd">        of `sl` `Residues`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">cas</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_reference_coords</span><span class="p">()</span>
        <span class="n">primitive_coords</span> <span class="o">=</span> <span class="n">make_primitive_extrapolate_ends</span><span class="p">(</span>
            <span class="n">cas</span><span class="p">,</span> <span class="n">smoothing_level</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sl</span><span class="p">)</span>
        <span class="n">primitive</span> <span class="o">=</span> <span class="n">Primitive</span><span class="o">.</span><span class="n">from_coordinates</span><span class="p">(</span><span class="n">primitive_coords</span><span class="p">)</span>
        <span class="n">primitive</span><span class="o">.</span><span class="n">relabel_monomers</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">])</span>
        <span class="n">primitive</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span>
        <span class="n">primitive</span><span class="o">.</span><span class="n">ampal_parent</span> <span class="o">=</span> <span class="bp">self</span>
        <span class="k">return</span> <span class="n">primitive</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">helices</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns a new `Assembly` containing only the alpha-helices.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Metadata is not currently preserved from the parent object.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        hel_assembly : Assembly</span>
<span class="sd">            `Assembly` containing only the alpha-helices of the</span>
<span class="sd">            original `Polymer`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tag_secondary_structure</span><span class="p">()</span>
        <span class="n">hel_assembly</span> <span class="o">=</span> <span class="n">find_ss_regions_polymer</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;H&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">hel_assembly</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">strands</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns a new `Assembly` containing only the beta-strand atoms.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Metadata is not currently preserved from the parent object.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        strand_assembly : Assembly</span>
<span class="sd">            `Assembly` containing only the beta-strand atoms of</span>
<span class="sd">            the original `Polymer`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tag_secondary_structure</span><span class="p">()</span>
        <span class="n">strand_assembly</span> <span class="o">=</span> <span class="n">find_ss_regions_polymer</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;E&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">strand_assembly</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">fasta</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Generates sequence data for the protein in FASTA format.&quot;&quot;&quot;</span>
        <span class="n">max_line_length</span> <span class="o">=</span> <span class="mi">79</span>
        <span class="n">fasta_str</span> <span class="o">=</span> <span class="s1">&#39;&gt;</span><span class="si">{0}</span><span class="s1">:</span><span class="si">{1}</span><span class="s1">|PDBID|CHAIN|SEQUENCE</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ampal_parent</span><span class="o">.</span><span class="n">id</span><span class="o">.</span><span class="n">upper</span><span class="p">(),</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span>
        <span class="n">split_seq</span> <span class="o">=</span> <span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="n">max_line_length</span><span class="p">]</span>
                     <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">),</span> <span class="n">max_line_length</span><span class="p">)]</span>
        <span class="k">for</span> <span class="n">seq_part</span> <span class="ow">in</span> <span class="n">split_seq</span><span class="p">:</span>
            <span class="n">fasta_str</span> <span class="o">+=</span> <span class="s1">&#39;</span><span class="si">{0}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">seq_part</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">fasta_str</span>

<div class="viewcode-block" id="Polypeptide.pack_new_sequence"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.pack_new_sequence">[docs]</a>    <span class="k">def</span> <span class="nf">pack_new_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Packs a new sequence onto the polymer using Scwrl4.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sequence : str</span>
<span class="sd">            String containing the amino acid sequence. This must</span>
<span class="sd">            be the same length as the Polymer</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ValueError</span>
<span class="sd">            Raised if the sequence length does not match the</span>
<span class="sd">            number of monomers in the Polymer.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># This import is here to prevent a circular import.</span>
        <span class="kn">from</span> <span class="nn">ampal.pdb_parser</span> <span class="k">import</span> <span class="n">convert_pdb_to_ampal</span>
        <span class="n">polymer_bb</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">backbone</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">polymer_bb</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s1">&#39;Sequence length (</span><span class="si">{}</span><span class="s1">) does not match Polymer length (</span><span class="si">{}</span><span class="s1">).&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">polymer_bb</span><span class="p">)))</span>
        <span class="n">scwrl_out</span> <span class="o">=</span> <span class="n">pack_sidechains</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">backbone</span><span class="o">.</span><span class="n">pdb</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">scwrl_out</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">packed_structure</span><span class="p">,</span> <span class="n">scwrl_score</span> <span class="o">=</span> <span class="n">scwrl_out</span>
        <span class="n">new_assembly</span> <span class="o">=</span> <span class="n">convert_pdb_to_ampal</span><span class="p">(</span><span class="n">packed_structure</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span> <span class="o">=</span> <span class="n">new_assembly</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[:]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;scwrl_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">scwrl_score</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assign_force_field</span><span class="p">(</span><span class="n">global_settings</span><span class="p">[</span><span class="s1">&#39;buff&#39;</span><span class="p">][</span><span class="s1">&#39;force_field&#39;</span><span class="p">])</span>
        <span class="k">return</span></div>

<div class="viewcode-block" id="Polypeptide.repack"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.repack">[docs]</a>    <span class="k">def</span> <span class="nf">repack</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Repacks current side chain sequence using Scwrl4.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pack_new_sequence</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
        <span class="k">return</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns the sequence of the `Polymer` as a string.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        sequence : str</span>
<span class="sd">            String of the `Residue` sequence of the `Polypeptide`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">mol_letter</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">]</span>
        <span class="k">return</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">molecular_weight</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns the molecular weight of the `Assembly` in Daltons.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">sequence_molecular_weight</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">molar_extinction_280</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns the extinction co-efficient of the `Assembly` at 280 nm.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">sequence_molar_extinction_280</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">isoelectric_point</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns the isoelectric point of the `Assembly`.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">sequence_isoelectric_point</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">backbone_bond_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Dictionary containing backbone bond lengths as lists of floats.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        bond_lengths : dict</span>
<span class="sd">            Keys are `n_ca`, `ca_c`, `c_o` and `c_n`, referring to the</span>
<span class="sd">            N-CA, CA-C, C=O and C-N bonds respectively. Values are</span>
<span class="sd">            lists of floats : the bond lengths in Angstroms.</span>
<span class="sd">            The lists of n_ca, ca_c and c_o are of length k for</span>
<span class="sd">            a Polypeptide containing k Residues. The list of c_n bonds</span>
<span class="sd">            is of length k-1 for a Polypeptide containing k Residues</span>
<span class="sd">            (C-N formed between successive `Residue` pairs).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">bond_lengths</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="n">n_ca</span><span class="o">=</span><span class="p">[</span><span class="n">distance</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">])</span>
                  <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_monomers</span><span class="p">(</span><span class="n">ligands</span><span class="o">=</span><span class="kc">False</span><span class="p">)],</span>
            <span class="n">ca_c</span><span class="o">=</span><span class="p">[</span><span class="n">distance</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">])</span>
                  <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_monomers</span><span class="p">(</span><span class="n">ligands</span><span class="o">=</span><span class="kc">False</span><span class="p">)],</span>
            <span class="n">c_o</span><span class="o">=</span><span class="p">[</span><span class="n">distance</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;O&#39;</span><span class="p">])</span>
                 <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_monomers</span><span class="p">(</span><span class="n">ligands</span><span class="o">=</span><span class="kc">False</span><span class="p">)],</span>
            <span class="n">c_n</span><span class="o">=</span><span class="p">[</span><span class="n">distance</span><span class="p">(</span><span class="n">r1</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">r2</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">])</span> <span class="k">for</span> <span class="n">r1</span><span class="p">,</span> <span class="n">r2</span> <span class="ow">in</span> <span class="p">[</span>
                <span class="p">(</span><span class="bp">self</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="bp">self</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)]],</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">bond_lengths</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">backbone_bond_angles</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Dictionary containing backbone bond angles as lists of floats.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        bond_angles : dict</span>
<span class="sd">            Keys are `n_ca_c`, `ca_c_o`, `ca_c_n` and `c_n_ca`, referring</span>
<span class="sd">            to the N-CA-C, CA-C=O, CA-C-N and C-N-CA angles respectively.</span>
<span class="sd">            Values are lists of floats : the bond angles in degrees.</span>
<span class="sd">            The lists of n_ca_c, ca_c_o are of length k for a `Polypeptide`</span>
<span class="sd">            containing k `Residues`. The list of ca_c_n and c_n_ca are of</span>
<span class="sd">            length k-1 for a `Polypeptide` containing k `Residues` (These</span>
<span class="sd">            angles are across the peptide bond, and are therefore formed</span>
<span class="sd">            between successive `Residue` pairs).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">bond_angles</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="n">n_ca_c</span><span class="o">=</span><span class="p">[</span><span class="n">angle_between_vectors</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">])</span>
                    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_monomers</span><span class="p">(</span><span class="n">ligands</span><span class="o">=</span><span class="kc">False</span><span class="p">)],</span>
            <span class="n">ca_c_o</span><span class="o">=</span><span class="p">[</span><span class="n">angle_between_vectors</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;O&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">])</span>
                    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_monomers</span><span class="p">(</span><span class="n">ligands</span><span class="o">=</span><span class="kc">False</span><span class="p">)],</span>
            <span class="n">ca_c_n</span><span class="o">=</span><span class="p">[</span><span class="n">angle_between_vectors</span><span class="p">(</span><span class="n">r1</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">r2</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">])</span>
                    <span class="k">for</span> <span class="n">r1</span><span class="p">,</span> <span class="n">r2</span> <span class="ow">in</span> <span class="p">[(</span><span class="bp">self</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="bp">self</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)]],</span>
            <span class="n">c_n_ca</span><span class="o">=</span><span class="p">[</span><span class="n">angle_between_vectors</span><span class="p">(</span><span class="n">r1</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">r2</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">],</span> <span class="n">r2</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">r2</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">])</span>
                    <span class="k">for</span> <span class="n">r1</span><span class="p">,</span> <span class="n">r2</span> <span class="ow">in</span> <span class="p">[(</span><span class="bp">self</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="bp">self</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)]],</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">bond_angles</span>

<div class="viewcode-block" id="Polypeptide.c_join"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.c_join">[docs]</a>    <span class="k">def</span> <span class="nf">c_join</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="n">psi</span><span class="o">=-</span><span class="mf">40.76</span><span class="p">,</span> <span class="n">omega</span><span class="o">=-</span><span class="mf">178.25</span><span class="p">,</span> <span class="n">phi</span><span class="o">=-</span><span class="mf">65.07</span><span class="p">,</span>
               <span class="n">o_c_n_angle</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">c_n_ca_angle</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">c_n_length</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
               <span class="n">relabel</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Joins other to self at the C-terminus via a peptide bond.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        This function directly modifies self. It does not return a new object.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        other: Residue or Polypeptide</span>
<span class="sd">        psi: float, optional</span>
<span class="sd">            Psi torsion angle (degrees) between final `Residue` of self</span>
<span class="sd">            and first `Residue` of other.</span>
<span class="sd">        omega: float, optional</span>
<span class="sd">            Omega torsion angle (degrees) between final `Residue` of</span>
<span class="sd">            self and first `Residue` of other.</span>
<span class="sd">        phi: float, optional</span>
<span class="sd">            Phi torsion angle (degrees) between final `Residue` of self</span>
<span class="sd">            and first `Residue` of other.</span>
<span class="sd">        o_c_n_angle: float or None, optional</span>
<span class="sd">            Desired angle between O, C (final `Residue` of self) and N</span>
<span class="sd">            (first `Residue` of other) atoms. If `None`, default value is</span>
<span class="sd">            taken from `ideal_backbone_bond_angles`.</span>
<span class="sd">        c_n_ca_angle: float or None, optional</span>
<span class="sd">            Desired angle between C (final `Residue` of self) and N, CA</span>
<span class="sd">            (first `Residue` of other) atoms. If `None`, default value is</span>
<span class="sd">            taken from `ideal_backbone_bond_angles`.</span>
<span class="sd">        c_n_length: float or None, optional</span>
<span class="sd">            Desired peptide bond length between final `Residue` of self</span>
<span class="sd">            and first `Residue` of other. If `None`, default value is taken</span>
<span class="sd">            from `ideal_backbone_bond_lengths`.</span>
<span class="sd">        relabel: bool, optional</span>
<span class="sd">            If `True`, `relabel_all` is run on self before returning.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError:</span>
<span class="sd">            If other is not a `Residue` or a Polypeptide.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">Residue</span><span class="p">):</span>
            <span class="n">other</span> <span class="o">=</span> <span class="n">Polypeptide</span><span class="p">([</span><span class="n">other</span><span class="p">])</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">Polypeptide</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
                <span class="s1">&#39;Only Polypeptide or Residue objects can be joined to a Polypeptide&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">omega</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">90</span><span class="p">:</span>
            <span class="n">peptide_conformation</span> <span class="o">=</span> <span class="s1">&#39;trans&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">peptide_conformation</span> <span class="o">=</span> <span class="s1">&#39;cis&#39;</span>
        <span class="k">if</span> <span class="n">o_c_n_angle</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">o_c_n_angle</span> <span class="o">=</span> <span class="n">ideal_backbone_bond_angles</span><span class="p">[</span><span class="n">peptide_conformation</span><span class="p">][</span><span class="s1">&#39;o_c_n&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">c_n_ca_angle</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">c_n_ca_angle</span> <span class="o">=</span> <span class="n">ideal_backbone_bond_angles</span><span class="p">[</span><span class="n">peptide_conformation</span><span class="p">][</span><span class="s1">&#39;c_n_ca&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">c_n_length</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">c_n_length</span> <span class="o">=</span> <span class="n">ideal_backbone_bond_lengths</span><span class="p">[</span><span class="s1">&#39;c_n&#39;</span><span class="p">]</span>
        <span class="n">r1</span> <span class="o">=</span> <span class="bp">self</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">r1_ca</span> <span class="o">=</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span>
        <span class="n">r1_c</span> <span class="o">=</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span>
        <span class="n">r1_o</span> <span class="o">=</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;O&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span>
        <span class="c1"># p1 is point that will be used to position the N atom of r2.</span>
        <span class="n">p1</span> <span class="o">=</span> <span class="n">r1_o</span><span class="p">[:]</span>
        <span class="c1"># rotate p1 by o_c_n_angle, about axis perpendicular to the</span>
        <span class="c1"># r1_ca, r1_c, r1_o plane, passing through r1_c.</span>
        <span class="n">axis</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">cross</span><span class="p">((</span><span class="n">r1_ca</span> <span class="o">-</span> <span class="n">r1_c</span><span class="p">),</span> <span class="p">(</span><span class="n">r1_o</span> <span class="o">-</span> <span class="n">r1_c</span><span class="p">))</span>
        <span class="n">q</span> <span class="o">=</span> <span class="n">Quaternion</span><span class="o">.</span><span class="n">angle_and_axis</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="n">o_c_n_angle</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">)</span>
        <span class="n">p1</span> <span class="o">=</span> <span class="n">q</span><span class="o">.</span><span class="n">rotate_vector</span><span class="p">(</span><span class="n">v</span><span class="o">=</span><span class="n">p1</span><span class="p">,</span> <span class="n">point</span><span class="o">=</span><span class="n">r1_c</span><span class="p">)</span>
        <span class="c1"># Ensure p1 is separated from r1_c by the correct distance.</span>
        <span class="n">p1</span> <span class="o">=</span> <span class="n">r1_c</span> <span class="o">+</span> <span class="p">(</span><span class="n">c_n_length</span> <span class="o">*</span> <span class="n">unit_vector</span><span class="p">(</span><span class="n">p1</span> <span class="o">-</span> <span class="n">r1_c</span><span class="p">))</span>
        <span class="c1"># rotate p1 and r1[&#39;O&#39;] by to obtain desired psi value at the join.</span>
        <span class="n">measured_psi</span> <span class="o">=</span> <span class="n">dihedral</span><span class="p">(</span><span class="n">r1</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">],</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">p1</span><span class="p">)</span>
        <span class="n">q</span> <span class="o">=</span> <span class="n">Quaternion</span><span class="o">.</span><span class="n">angle_and_axis</span><span class="p">(</span>
            <span class="n">angle</span><span class="o">=</span><span class="p">(</span><span class="n">psi</span> <span class="o">-</span> <span class="n">measured_psi</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="n">r1_c</span> <span class="o">-</span> <span class="n">r1_ca</span><span class="p">))</span>
        <span class="n">p1</span> <span class="o">=</span> <span class="n">q</span><span class="o">.</span><span class="n">rotate_vector</span><span class="p">(</span><span class="n">v</span><span class="o">=</span><span class="n">p1</span><span class="p">,</span> <span class="n">point</span><span class="o">=</span><span class="n">r1_c</span><span class="p">)</span>
        <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;O&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span> <span class="o">=</span> <span class="n">q</span><span class="o">.</span><span class="n">rotate_vector</span><span class="p">(</span><span class="n">v</span><span class="o">=</span><span class="n">r1_o</span><span class="p">,</span> <span class="n">point</span><span class="o">=</span><span class="n">r1_c</span><span class="p">)</span>
        <span class="c1"># translate other so that its first N atom is at p1</span>
        <span class="n">other</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">vector</span><span class="o">=</span><span class="p">(</span><span class="n">p1</span> <span class="o">-</span> <span class="n">other</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">))</span>
        <span class="c1"># rotate other so that c_n_ca angle is correct.</span>
        <span class="n">v1</span> <span class="o">=</span> <span class="n">r1_c</span> <span class="o">-</span> <span class="n">other</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span>
        <span class="n">v2</span> <span class="o">=</span> <span class="n">other</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span> <span class="o">-</span> <span class="n">other</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span>
        <span class="n">measured_c_n_ca</span> <span class="o">=</span> <span class="n">angle_between_vectors</span><span class="p">(</span><span class="n">v1</span><span class="p">,</span> <span class="n">v2</span><span class="p">)</span>
        <span class="n">axis</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">cross</span><span class="p">(</span><span class="n">v1</span><span class="p">,</span> <span class="n">v2</span><span class="p">)</span>
        <span class="n">other</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="p">(</span><span class="n">c_n_ca_angle</span> <span class="o">-</span> <span class="n">measured_c_n_ca</span><span class="p">),</span>
                     <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">,</span> <span class="n">point</span><span class="o">=</span><span class="n">other</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">)</span>
        <span class="c1"># rotate other to obtain desired omega and phi values at the join</span>
        <span class="n">measured_omega</span> <span class="o">=</span> <span class="n">dihedral</span><span class="p">(</span>
            <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">other</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">],</span> <span class="n">other</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">])</span>
        <span class="n">other</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="p">(</span><span class="n">omega</span> <span class="o">-</span> <span class="n">measured_omega</span><span class="p">),</span>
                     <span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="n">other</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]),</span> <span class="n">point</span><span class="o">=</span><span class="n">other</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">)</span>
        <span class="n">measured_phi</span> <span class="o">=</span> <span class="n">dihedral</span><span class="p">(</span>
            <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">other</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">],</span> <span class="n">other</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span> <span class="n">other</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">])</span>
        <span class="n">other</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="p">(</span><span class="n">phi</span> <span class="o">-</span> <span class="n">measured_phi</span><span class="p">),</span>
                     <span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="n">other</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">other</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">]),</span> <span class="n">point</span><span class="o">=</span><span class="n">other</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">other</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">relabel</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">relabel_all</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;assigned_ff&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">return</span></div>

<div class="viewcode-block" id="Polypeptide.n_join"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.n_join">[docs]</a>    <span class="k">def</span> <span class="nf">n_join</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="n">psi</span><span class="o">=-</span><span class="mf">40.76</span><span class="p">,</span> <span class="n">omega</span><span class="o">=-</span><span class="mf">178.25</span><span class="p">,</span> <span class="n">phi</span><span class="o">=-</span><span class="mf">65.07</span><span class="p">,</span>
               <span class="n">o_c_n_angle</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">c_n_ca_angle</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">c_n_length</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">relabel</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Joins other to self at the N-terminus via a peptide bond.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        This function directly modifies self. It does not return a new object.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        other: Residue or Polypeptide</span>
<span class="sd">        psi: float</span>
<span class="sd">            Psi torsion angle (degrees) between final `Residue` of other</span>
<span class="sd">            and first `Residue` of self.</span>
<span class="sd">        omega: float</span>
<span class="sd">            Omega torsion angle (degrees) between final `Residue` of</span>
<span class="sd">            other and first `Residue` of self.</span>
<span class="sd">        phi: float</span>
<span class="sd">            Phi torsion angle (degrees) between final `Residue` of other</span>
<span class="sd">            and first `Residue` of self.</span>
<span class="sd">        o_c_n_angle: float or None</span>
<span class="sd">            Desired angle between O, C (final `Residue` of other) and N</span>
<span class="sd">            (first `Residue` of self) atoms. If `None`, default value is</span>
<span class="sd">            taken from `ideal_backbone_bond_angles`.</span>
<span class="sd">        c_n_ca_angle: float or None</span>
<span class="sd">            Desired angle between C (final `Residue` of other) and N, CA</span>
<span class="sd">            (first `Residue` of self) atoms. If `None`, default value is taken</span>
<span class="sd">            from `ideal_backbone_bond_angles`.</span>
<span class="sd">        c_n_length: float or None</span>
<span class="sd">            Desired peptide bond length between final `Residue` of other</span>
<span class="sd">            and first `Residue` of self. If None, default value is taken</span>
<span class="sd">            from ideal_backbone_bond_lengths.</span>
<span class="sd">        relabel: bool</span>
<span class="sd">            If True, relabel_all is run on self before returning.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError:</span>
<span class="sd">            If other is not a `Residue` or a `Polypeptide`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">Residue</span><span class="p">):</span>
            <span class="n">other</span> <span class="o">=</span> <span class="n">Polypeptide</span><span class="p">([</span><span class="n">other</span><span class="p">])</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">Polypeptide</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
                <span class="s1">&#39;Only Polypeptide or Residue objects can be joined to a Polypeptide&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">omega</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">90</span><span class="p">:</span>
            <span class="n">peptide_conformation</span> <span class="o">=</span> <span class="s1">&#39;trans&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">peptide_conformation</span> <span class="o">=</span> <span class="s1">&#39;cis&#39;</span>
        <span class="k">if</span> <span class="n">o_c_n_angle</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">o_c_n_angle</span> <span class="o">=</span> <span class="n">ideal_backbone_bond_angles</span><span class="p">[</span><span class="n">peptide_conformation</span><span class="p">][</span><span class="s1">&#39;o_c_n&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">c_n_ca_angle</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">c_n_ca_angle</span> <span class="o">=</span> <span class="n">ideal_backbone_bond_angles</span><span class="p">[</span><span class="n">peptide_conformation</span><span class="p">][</span><span class="s1">&#39;c_n_ca&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">c_n_length</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">c_n_length</span> <span class="o">=</span> <span class="n">ideal_backbone_bond_lengths</span><span class="p">[</span><span class="s1">&#39;c_n&#39;</span><span class="p">]</span>
        <span class="n">r1</span> <span class="o">=</span> <span class="bp">self</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">r1_n</span> <span class="o">=</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span>
        <span class="n">r1_ca</span> <span class="o">=</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span>
        <span class="n">r1_c</span> <span class="o">=</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span>
        <span class="c1"># p1 is point that will be used to position the C atom of r2.</span>
        <span class="n">p1</span> <span class="o">=</span> <span class="n">r1_ca</span><span class="p">[:]</span>
        <span class="c1"># rotate p1 by c_n_ca_angle, about axis perpendicular to the</span>
        <span class="c1"># r1_n, r1_ca, r1_c plane, passing through r1_ca.</span>
        <span class="n">axis</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">cross</span><span class="p">((</span><span class="n">r1_ca</span> <span class="o">-</span> <span class="n">r1_n</span><span class="p">),</span> <span class="p">(</span><span class="n">r1_c</span> <span class="o">-</span> <span class="n">r1_n</span><span class="p">))</span>
        <span class="n">q</span> <span class="o">=</span> <span class="n">Quaternion</span><span class="o">.</span><span class="n">angle_and_axis</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="n">c_n_ca_angle</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">)</span>
        <span class="n">p1</span> <span class="o">=</span> <span class="n">q</span><span class="o">.</span><span class="n">rotate_vector</span><span class="p">(</span><span class="n">v</span><span class="o">=</span><span class="n">p1</span><span class="p">,</span> <span class="n">point</span><span class="o">=</span><span class="n">r1_n</span><span class="p">)</span>
        <span class="c1"># Ensure p1 is separated from r1_n by the correct distance.</span>
        <span class="n">p1</span> <span class="o">=</span> <span class="n">r1_n</span> <span class="o">+</span> <span class="p">(</span><span class="n">c_n_length</span> <span class="o">*</span> <span class="n">unit_vector</span><span class="p">(</span><span class="n">p1</span> <span class="o">-</span> <span class="n">r1_n</span><span class="p">))</span>
        <span class="c1"># translate other so that its final C atom is at p1</span>
        <span class="n">other</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">vector</span><span class="o">=</span><span class="p">(</span><span class="n">p1</span> <span class="o">-</span> <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">))</span>
        <span class="c1"># Force CA-C=O-N to be in a plane, and fix O=C-N angle accordingly</span>
        <span class="n">measured_dihedral</span> <span class="o">=</span> <span class="n">dihedral</span><span class="p">(</span>
            <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span> <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;O&#39;</span><span class="p">],</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">])</span>
        <span class="n">desired_dihedral</span> <span class="o">=</span> <span class="mf">180.0</span>
        <span class="n">axis</span> <span class="o">=</span> <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;O&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">]</span>
        <span class="n">other</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="p">(</span><span class="n">measured_dihedral</span> <span class="o">-</span> <span class="n">desired_dihedral</span><span class="p">),</span>
                     <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">,</span> <span class="n">point</span><span class="o">=</span><span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">)</span>
        <span class="n">axis</span> <span class="o">=</span> <span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">cross</span><span class="p">(</span><span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;O&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                            <span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">]))</span>
        <span class="n">measured_o_c_n</span> <span class="o">=</span> <span class="n">angle_between_vectors</span><span class="p">(</span>
            <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;O&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">])</span>
        <span class="n">other</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="p">(</span><span class="n">measured_o_c_n</span> <span class="o">-</span> <span class="n">o_c_n_angle</span><span class="p">),</span>
                     <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">,</span> <span class="n">point</span><span class="o">=</span><span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">)</span>
        <span class="c1"># rotate other to obtain desired phi, omega, psi values at the join.</span>
        <span class="n">measured_phi</span> <span class="o">=</span> <span class="n">dihedral</span><span class="p">(</span><span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">],</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">])</span>
        <span class="n">other</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="p">(</span><span class="n">phi</span> <span class="o">-</span> <span class="n">measured_phi</span><span class="p">),</span>
                     <span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="n">r1_n</span> <span class="o">-</span> <span class="n">r1_ca</span><span class="p">),</span> <span class="n">point</span><span class="o">=</span><span class="n">r1_ca</span><span class="p">)</span>
        <span class="n">measured_omega</span> <span class="o">=</span> <span class="n">dihedral</span><span class="p">(</span>
            <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span> <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">],</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">])</span>
        <span class="n">other</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="p">(</span><span class="n">measured_omega</span> <span class="o">-</span> <span class="n">omega</span><span class="p">),</span>
                     <span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="n">r1</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">]),</span> <span class="n">point</span><span class="o">=</span><span class="n">r1_n</span><span class="p">)</span>
        <span class="n">measured_psi</span> <span class="o">=</span> <span class="n">dihedral</span><span class="p">(</span>
            <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">],</span> <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span> <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">r1</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">])</span>
        <span class="n">other</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">angle</span><span class="o">=-</span><span class="p">(</span><span class="n">measured_psi</span> <span class="o">-</span> <span class="n">psi</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">]),</span>
                     <span class="n">point</span><span class="o">=</span><span class="n">other</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">_monomers</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span>
        <span class="k">if</span> <span class="n">relabel</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">relabel_all</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;assigned_ff&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">return</span></div>

<div class="viewcode-block" id="Polypeptide.tag_secondary_structure"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.tag_secondary_structure">[docs]</a>    <span class="k">def</span> <span class="nf">tag_secondary_structure</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Tags each `Residue` of the `Polypeptide` with secondary structure.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        DSSP must be available to call. Check by running</span>
<span class="sd">        `isambard.external_programs.dssp.test_dssp`. If DSSP is not</span>
<span class="sd">        available, please follow instruction here to add it:</span>
<span class="sd">        https://github.com/woolfson-group/isambard#external-programs</span>

<span class="sd">        For more information on DSSP see [1].</span>

<span class="sd">        References</span>
<span class="sd">        ----------</span>
<span class="sd">        .. [1] Kabsch W, Sander C (1983) &quot;Dictionary of protein</span>
<span class="sd">           secondary structure: pattern recognition of hydrogen-bonded</span>
<span class="sd">           and geometrical features&quot;, Biopolymers, 22, 2577-637.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        force : bool, optional</span>
<span class="sd">            If `True` the tag will be run even if `Residues` are</span>
<span class="sd">            already tagged.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">tagged</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;secondary_structure&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">tags</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
                  <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">]</span>
        <span class="k">if</span> <span class="p">(</span><span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">tagged</span><span class="p">))</span> <span class="ow">or</span> <span class="n">force</span><span class="p">:</span>
            <span class="n">dssp_out</span> <span class="o">=</span> <span class="n">run_dssp</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pdb</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">dssp_out</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">return</span>
            <span class="n">dssp_ss_list</span> <span class="o">=</span> <span class="n">extract_all_ss_dssp</span><span class="p">(</span><span class="n">dssp_out</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">monomer</span><span class="p">,</span> <span class="n">dssp_ss</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">,</span> <span class="n">dssp_ss_list</span><span class="p">):</span>
                <span class="n">monomer</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;secondary_structure&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dssp_ss</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">return</span></div>

<div class="viewcode-block" id="Polypeptide.tag_residue_solvent_accessibility"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.tag_residue_solvent_accessibility">[docs]</a>    <span class="k">def</span> <span class="nf">tag_residue_solvent_accessibility</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tag_type</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">tag_total</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                          <span class="n">force</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">include_hetatms</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Tags `Residues` wirh relative residue solvent accessibility.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        THIS FUNCTIONALITY REQUIRES NACESS.</span>
<span class="sd">        This function tags the Monomer with the *relative* RSA of</span>
<span class="sd">        the *whole side chain*, i.e. column 2 of the .rsa file that</span>
<span class="sd">        NACCESS writes.</span>

<span class="sd">        References</span>
<span class="sd">        ----------</span>
<span class="sd">        .. [1] Hubbard,S.J. &amp; Thornton, J.M. (1993), &#39;NACCESS&#39;,</span>
<span class="sd">           Computer Program, Department of Biochemistry and Molecular</span>
<span class="sd">           Biology, University College London.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        force : bool, optional</span>
<span class="sd">            If `True`, the ta will be run even if `Residues` are</span>
<span class="sd">            already tagged.</span>
<span class="sd">        tag_type : str, optional</span>
<span class="sd">            Specifies the name of the tag. Defaults to</span>
<span class="sd">            &#39;residue_solvent_accessibility&#39;. Useful for specifying more</span>
<span class="sd">            than one tag, e.g. if the Polymer is part of an Assembly.</span>
<span class="sd">        tag_total : bool, optional</span>
<span class="sd">            If True then the total rsa of the Polymer will be tagged</span>
<span class="sd">            in the &#39;total accessibility&#39; tag.</span>
<span class="sd">        include_hetatms:bool, optional</span>
<span class="sd">            If true then NACCESS will run with the -h flag and will</span>
<span class="sd">            include heteroatom solvent accessibility where it can.</span>
<span class="sd">            Helpful if your file has MSE residues that you don&#39;t</span>
<span class="sd">            convert to MET, but best check if they are there</span>
<span class="sd">            before using the flag.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">tag_type</span><span class="p">:</span>
            <span class="n">tag_type</span> <span class="o">=</span> <span class="n">tag_type</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">tag_type</span> <span class="o">=</span> <span class="s1">&#39;residue_solvent_accessibility&#39;</span>
        <span class="n">tagged</span> <span class="o">=</span> <span class="p">[</span><span class="n">tag_type</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">tags</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">]</span>
        <span class="k">if</span> <span class="p">(</span><span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">tagged</span><span class="p">))</span> <span class="ow">or</span> <span class="n">force</span><span class="p">:</span>
            <span class="n">naccess_rsa_list</span><span class="p">,</span> <span class="n">total</span> <span class="o">=</span> <span class="n">extract_residue_accessibility</span><span class="p">(</span><span class="n">run_naccess</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">pdb</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;rsa&#39;</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">include_hetatms</span><span class="o">=</span><span class="n">include_hetatms</span><span class="p">),</span> <span class="n">path</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">get_total</span><span class="o">=</span><span class="n">tag_total</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">monomer</span><span class="p">,</span> <span class="n">naccess_rsa</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">,</span> <span class="n">naccess_rsa_list</span><span class="p">):</span>
                <span class="n">monomer</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="n">tag_type</span><span class="p">]</span> <span class="o">=</span> <span class="n">naccess_rsa</span>
            <span class="k">if</span> <span class="n">tag_total</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;total_polymer_accessibility&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">total</span>
        <span class="k">return</span></div>

<div class="viewcode-block" id="Polypeptide.tag_dssp_solvent_accessibility"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.tag_dssp_solvent_accessibility">[docs]</a>    <span class="k">def</span> <span class="nf">tag_dssp_solvent_accessibility</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Tags each `Residues` Polymer with its solvent accessibility.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        For more about DSSP&#39;s solvent accessibilty metric, see:</span>
<span class="sd">            http://swift.cmbi.ru.nl/gv/dssp/HTML/descrip.html#ACC</span>

<span class="sd">        References</span>
<span class="sd">        ----------</span>
<span class="sd">        .. [1] Kabsch W, Sander C (1983) &quot;Dictionary of protein</span>
<span class="sd">           secondary structure: pattern recognition of hydrogen-bonded</span>
<span class="sd">           and geometrical features&quot;, Biopolymers, 22, 2577-637.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        force : bool, optional</span>
<span class="sd">            If `True` the tag will be run even if `Residues` are</span>
<span class="sd">            already tagged.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">tagged</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;dssp_acc&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">tags</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">]</span>
        <span class="k">if</span> <span class="p">(</span><span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">tagged</span><span class="p">))</span> <span class="ow">or</span> <span class="n">force</span><span class="p">:</span>
            <span class="n">dssp_out</span> <span class="o">=</span> <span class="n">run_dssp</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pdb</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">dssp_out</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">return</span>
            <span class="n">dssp_acc_list</span> <span class="o">=</span> <span class="n">extract_solvent_accessibility_dssp</span><span class="p">(</span>
                <span class="n">dssp_out</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">monomer</span><span class="p">,</span> <span class="n">dssp_acc</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">,</span> <span class="n">dssp_acc_list</span><span class="p">):</span>
                <span class="n">monomer</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;dssp_acc&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dssp_acc</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">return</span></div>

<div class="viewcode-block" id="Polypeptide.tag_sidechain_dihedrals"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.tag_sidechain_dihedrals">[docs]</a>    <span class="k">def</span> <span class="nf">tag_sidechain_dihedrals</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Tags each monomer with side-chain dihedral angles</span>

<span class="sd">        force: bool, optional</span>
<span class="sd">            If `True` the tag will be run even if `Residues` are</span>
<span class="sd">            already tagged.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">tagged</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;chi_angles&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">tags</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">]</span>
        <span class="k">if</span> <span class="p">(</span><span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">tagged</span><span class="p">))</span> <span class="ow">or</span> <span class="n">force</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">monomer</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">:</span>
                <span class="n">chi_angles</span> <span class="o">=</span> <span class="n">measure_sidechain_torsion_angles</span><span class="p">(</span>
                    <span class="n">monomer</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">monomer</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;chi_angles&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">chi_angles</span>
        <span class="k">return</span></div>

<div class="viewcode-block" id="Polypeptide.tag_torsion_angles"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.tag_torsion_angles">[docs]</a>    <span class="k">def</span> <span class="nf">tag_torsion_angles</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Tags each Monomer of the Polymer with its omega, phi and psi torsion angle.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        force : bool, optional</span>
<span class="sd">            If `True` the tag will be run even if `Residues` are</span>
<span class="sd">            already tagged.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">tagged</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;omega&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">tags</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">]</span>
        <span class="k">if</span> <span class="p">(</span><span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">tagged</span><span class="p">))</span> <span class="ow">or</span> <span class="n">force</span><span class="p">:</span>
            <span class="n">tas</span> <span class="o">=</span> <span class="n">measure_torsion_angles</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">monomer</span><span class="p">,</span> <span class="p">(</span><span class="n">omega</span><span class="p">,</span> <span class="n">phi</span><span class="p">,</span> <span class="n">psi</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">,</span> <span class="n">tas</span><span class="p">):</span>
                <span class="n">monomer</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;omega&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">omega</span>
                <span class="n">monomer</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;phi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">phi</span>
                <span class="n">monomer</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;psi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">psi</span>
                <span class="n">monomer</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;tas&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">omega</span><span class="p">,</span> <span class="n">phi</span><span class="p">,</span> <span class="n">psi</span><span class="p">)</span>
        <span class="k">return</span></div>

<div class="viewcode-block" id="Polypeptide.rise_per_residue"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.rise_per_residue">[docs]</a>    <span class="k">def</span> <span class="nf">rise_per_residue</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;List of rise per residue values along the `Polypeptide`.</span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Calculated from `Polypeptide.primitive`.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">primitive</span><span class="o">.</span><span class="n">rise_per_residue</span><span class="p">()</span></div>

<div class="viewcode-block" id="Polypeptide.radii_of_curvature"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.radii_of_curvature">[docs]</a>    <span class="k">def</span> <span class="nf">radii_of_curvature</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; List of radius of curvature values along the `Polypeptide`.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">primitive</span><span class="o">.</span><span class="n">radii_of_curvature</span><span class="p">()</span></div>

<div class="viewcode-block" id="Polypeptide.tag_ca_geometry"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.tag_ca_geometry">[docs]</a>    <span class="k">def</span> <span class="nf">tag_ca_geometry</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">reference_axis</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                        <span class="n">reference_axis_name</span><span class="o">=</span><span class="s1">&#39;ref_axis&#39;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Tags each `Residue` with rise_per_residue, radius_of_curvature and residues_per_turn.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        force : bool, optional</span>
<span class="sd">            If `True` the tag will be run even if `Residues` are already</span>
<span class="sd">            tagged.</span>
<span class="sd">        reference_axis : list(numpy.array or tuple or list), optional</span>
<span class="sd">            Coordinates to feed to geometry functions that depend on</span>
<span class="sd">            having a reference axis.</span>
<span class="sd">        reference_axis_name : str, optional</span>
<span class="sd">            Used to name the keys in tags at `Polypeptide` and `Residue` level.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">tagged</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;rise_per_residue&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">tags</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">]</span>
        <span class="k">if</span> <span class="p">(</span><span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">tagged</span><span class="p">))</span> <span class="ow">or</span> <span class="n">force</span><span class="p">:</span>
            <span class="c1"># Assign tags None if Polymer is too short to have a primitive.</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">7</span><span class="p">:</span>
                <span class="n">rprs</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
                <span class="n">rocs</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
                <span class="n">rpts</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">rprs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">rise_per_residue</span><span class="p">()</span>
                <span class="n">rocs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">radii_of_curvature</span><span class="p">()</span>
                <span class="n">rpts</span> <span class="o">=</span> <span class="n">residues_per_turn</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">monomer</span><span class="p">,</span> <span class="n">rpr</span><span class="p">,</span> <span class="n">roc</span><span class="p">,</span> <span class="n">rpt</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">,</span> <span class="n">rprs</span><span class="p">,</span> <span class="n">rocs</span><span class="p">,</span> <span class="n">rpts</span><span class="p">):</span>
                <span class="n">monomer</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;rise_per_residue&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rpr</span>
                <span class="n">monomer</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;radius_of_curvature&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">roc</span>
                <span class="n">monomer</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;residues_per_turn&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rpt</span>
        <span class="c1"># Functions that require a reference_axis.</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">reference_axis</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">reference_axis</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)):</span>
            <span class="c1"># Set up arguments to pass to functions.</span>
            <span class="n">ref_axis_args</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">p</span><span class="o">=</span><span class="bp">self</span><span class="p">,</span>
                                 <span class="n">reference_axis</span><span class="o">=</span><span class="n">reference_axis</span><span class="p">,</span>
                                 <span class="n">tag</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                 <span class="n">reference_axis_name</span><span class="o">=</span><span class="n">reference_axis_name</span><span class="p">)</span>
            <span class="c1"># Run the functions.</span>
            <span class="n">polymer_to_reference_axis_distances</span><span class="p">(</span><span class="o">**</span><span class="n">ref_axis_args</span><span class="p">)</span>
            <span class="n">crick_angles</span><span class="p">(</span><span class="o">**</span><span class="n">ref_axis_args</span><span class="p">)</span>
            <span class="n">alpha_angles</span><span class="p">(</span><span class="o">**</span><span class="n">ref_axis_args</span><span class="p">)</span>
        <span class="k">return</span></div>

<div class="viewcode-block" id="Polypeptide.valid_backbone_bond_lengths"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.valid_backbone_bond_lengths">[docs]</a>    <span class="k">def</span> <span class="nf">valid_backbone_bond_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">atol</span><span class="o">=</span><span class="mf">0.1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;True if all backbone bonds are within atol Angstroms of the expected distance.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Ideal bond lengths taken from [1].</span>

<span class="sd">        References</span>
<span class="sd">        ----------</span>
<span class="sd">        .. [1] Schulz, G. E, and R. Heiner Schirmer. Principles Of</span>
<span class="sd">           Protein Structure. New York: Springer-Verlag, 1979.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        atol : float, optional</span>
<span class="sd">            Tolerance value in Angstoms for the absolute deviation</span>
<span class="sd">            away from ideal backbone bond lengths.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">bond_lengths</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">backbone_bond_lengths</span>
        <span class="n">a1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">bond_lengths</span><span class="p">[</span><span class="s1">&#39;n_ca&#39;</span><span class="p">],</span>
                            <span class="p">[</span><span class="n">ideal_backbone_bond_lengths</span><span class="p">[</span><span class="s1">&#39;n_ca&#39;</span><span class="p">]]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">),</span>
                            <span class="n">atol</span><span class="o">=</span><span class="n">atol</span><span class="p">)</span>
        <span class="n">a2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">bond_lengths</span><span class="p">[</span><span class="s1">&#39;ca_c&#39;</span><span class="p">],</span>
                            <span class="p">[</span><span class="n">ideal_backbone_bond_lengths</span><span class="p">[</span><span class="s1">&#39;ca_c&#39;</span><span class="p">]]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">),</span>
                            <span class="n">atol</span><span class="o">=</span><span class="n">atol</span><span class="p">)</span>
        <span class="n">a3</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">bond_lengths</span><span class="p">[</span><span class="s1">&#39;c_o&#39;</span><span class="p">],</span>
                            <span class="p">[</span><span class="n">ideal_backbone_bond_lengths</span><span class="p">[</span><span class="s1">&#39;c_o&#39;</span><span class="p">]]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">),</span>
                            <span class="n">atol</span><span class="o">=</span><span class="n">atol</span><span class="p">)</span>
        <span class="n">a4</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">bond_lengths</span><span class="p">[</span><span class="s1">&#39;c_n&#39;</span><span class="p">],</span>
                            <span class="p">[</span><span class="n">ideal_backbone_bond_lengths</span><span class="p">[</span><span class="s1">&#39;c_n&#39;</span><span class="p">]]</span> <span class="o">*</span>
                            <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">),</span>
                            <span class="n">atol</span><span class="o">=</span><span class="n">atol</span><span class="p">)</span>
        <span class="k">return</span> <span class="nb">all</span><span class="p">([</span><span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">,</span> <span class="n">a3</span><span class="p">,</span> <span class="n">a4</span><span class="p">])</span></div>

<div class="viewcode-block" id="Polypeptide.valid_backbone_bond_angles"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.valid_backbone_bond_angles">[docs]</a>    <span class="k">def</span> <span class="nf">valid_backbone_bond_angles</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">atol</span><span class="o">=</span><span class="mi">20</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;True if all backbone bond angles are within atol degrees of their expected values.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Ideal bond angles taken from [1].</span>

<span class="sd">        References</span>
<span class="sd">        ----------</span>
<span class="sd">        .. [1] Schulz, G. E, and R. Heiner Schirmer. Principles Of</span>
<span class="sd">           Protein Structure. New York: Springer-Verlag, 1979.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        atol : float, optional</span>
<span class="sd">            Tolerance value in degrees for the absolute deviation</span>
<span class="sd">            away from ideal backbone bond angles.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">bond_angles</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">backbone_bond_angles</span>
        <span class="n">omegas</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">measure_torsion_angles</span><span class="p">(</span><span class="bp">self</span><span class="p">)]</span>
        <span class="n">trans</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;trans&#39;</span> <span class="k">if</span> <span class="p">(</span><span class="n">omega</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span>
            <span class="nb">abs</span><span class="p">(</span><span class="n">omega</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">90</span><span class="p">)</span> <span class="k">else</span> <span class="s1">&#39;cis&#39;</span> <span class="k">for</span> <span class="n">omega</span> <span class="ow">in</span> <span class="n">omegas</span><span class="p">]</span>
        <span class="n">ideal_n_ca_c</span> <span class="o">=</span> <span class="p">[</span><span class="n">ideal_backbone_bond_angles</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="s1">&#39;n_ca_c&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">trans</span><span class="p">]</span>
        <span class="n">ideal_ca_c_o</span> <span class="o">=</span> <span class="p">[</span><span class="n">ideal_backbone_bond_angles</span><span class="p">[</span><span class="n">trans</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]]</span>
                        <span class="p">[</span><span class="s1">&#39;ca_c_o&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">trans</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)]</span>
        <span class="n">ideal_ca_c_o</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ideal_backbone_bond_angles</span><span class="p">[</span><span class="s1">&#39;trans&#39;</span><span class="p">][</span><span class="s1">&#39;ca_c_o&#39;</span><span class="p">])</span>
        <span class="n">ideal_ca_c_n</span> <span class="o">=</span> <span class="p">[</span><span class="n">ideal_backbone_bond_angles</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="s1">&#39;ca_c_n&#39;</span><span class="p">]</span>
                        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">trans</span><span class="p">[</span><span class="mi">1</span><span class="p">:]]</span>
        <span class="n">ideal_c_n_ca</span> <span class="o">=</span> <span class="p">[</span><span class="n">ideal_backbone_bond_angles</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="s1">&#39;c_n_ca&#39;</span><span class="p">]</span>
                        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">trans</span><span class="p">[</span><span class="mi">1</span><span class="p">:]]</span>
        <span class="n">a1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">bond_angles</span><span class="p">[</span><span class="s1">&#39;n_ca_c&#39;</span><span class="p">],</span> <span class="p">[</span><span class="n">ideal_n_ca_c</span><span class="p">],</span> <span class="n">atol</span><span class="o">=</span><span class="n">atol</span><span class="p">)</span>
        <span class="n">a2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">bond_angles</span><span class="p">[</span><span class="s1">&#39;ca_c_o&#39;</span><span class="p">],</span> <span class="p">[</span><span class="n">ideal_ca_c_o</span><span class="p">],</span> <span class="n">atol</span><span class="o">=</span><span class="n">atol</span><span class="p">)</span>
        <span class="n">a3</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">bond_angles</span><span class="p">[</span><span class="s1">&#39;ca_c_n&#39;</span><span class="p">],</span> <span class="p">[</span><span class="n">ideal_ca_c_n</span><span class="p">],</span> <span class="n">atol</span><span class="o">=</span><span class="n">atol</span><span class="p">)</span>
        <span class="n">a4</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">bond_angles</span><span class="p">[</span><span class="s1">&#39;c_n_ca&#39;</span><span class="p">],</span> <span class="p">[</span><span class="n">ideal_c_n_ca</span><span class="p">],</span> <span class="n">atol</span><span class="o">=</span><span class="n">atol</span><span class="p">)</span>
        <span class="k">return</span> <span class="nb">all</span><span class="p">([</span><span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">,</span> <span class="n">a3</span><span class="p">,</span> <span class="n">a4</span><span class="p">])</span></div>

<div class="viewcode-block" id="Polypeptide.c_cap"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.c_cap">[docs]</a>    <span class="k">def</span> <span class="nf">c_cap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cap</span><span class="o">=</span><span class="s1">&#39;acid&#39;</span><span class="p">,</span> <span class="n">cap_dihedral</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Caps C-terminus of polypeptide chain.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Default behaviour is to add an oxygen atom to create a</span>
<span class="sd">        carboxylate function at the C-terminus without changing the</span>
<span class="sd">        psi angle of the C-terminal residue. Alternative psi angles</span>
<span class="sd">        can be accessed through the cap_dihedral parameter. Will not</span>
<span class="sd">        remove an existing cap if one is present, though altering a</span>
<span class="sd">        cap of the same type will overwrite the original one.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        cap : str, optional</span>
<span class="sd">            Type of cap to be added. Options: &#39;acid&#39;, &#39;amide&#39;</span>
<span class="sd">        cap_dihedral : bool</span>
<span class="sd">            Alternate psi angle to be used when added cap.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">cap</span> <span class="o">==</span> <span class="s1">&#39;acid&#39;</span><span class="p">:</span>
            <span class="n">acetate</span> <span class="o">=</span> <span class="n">Ligand</span><span class="p">(</span><span class="n">atoms</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mol_code</span><span class="o">=</span><span class="s1">&#39;UNK&#39;</span><span class="p">,</span>
                             <span class="n">is_hetero</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ampal_parent</span><span class="o">=</span><span class="n">Polypeptide</span><span class="p">)</span>
            <span class="n">atoms</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Atom</span><span class="p">([</span><span class="o">-</span><span class="mf">1.4210</span><span class="p">,</span> <span class="mf">0.4120</span><span class="p">,</span> <span class="mf">0.0000</span><span class="p">],</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span>
                               <span class="n">res_label</span><span class="o">=</span><span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="n">ampal_parent</span><span class="o">=</span><span class="n">Ligand</span><span class="p">)</span>
            <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Atom</span><span class="p">([</span><span class="mf">0.0120</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.0560</span><span class="p">,</span> <span class="mf">0.0020</span><span class="p">],</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span>
                              <span class="n">res_label</span><span class="o">=</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">ampal_parent</span><span class="o">=</span><span class="n">Ligand</span><span class="p">)</span>
            <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;O&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Atom</span><span class="p">([</span><span class="mf">0.2610</span><span class="p">,</span> <span class="o">-</span><span class="mf">1.2380</span><span class="p">,</span> <span class="mf">0.0000</span><span class="p">],</span> <span class="s1">&#39;O&#39;</span><span class="p">,</span>
                              <span class="n">res_label</span><span class="o">=</span><span class="s1">&#39;O&#39;</span><span class="p">,</span> <span class="n">ampal_parent</span><span class="o">=</span><span class="n">Ligand</span><span class="p">)</span>
            <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;OXT&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Atom</span><span class="p">([</span><span class="mf">1.0110</span><span class="p">,</span> <span class="mf">0.8400</span><span class="p">,</span> <span class="mf">0.0000</span><span class="p">],</span>
                                <span class="s1">&#39;O&#39;</span><span class="p">,</span> <span class="n">res_label</span><span class="o">=</span><span class="s1">&#39;OXT&#39;</span><span class="p">,</span> <span class="n">ampal_parent</span><span class="o">=</span><span class="n">Ligand</span><span class="p">)</span>
            <span class="n">acetate</span><span class="o">.</span><span class="n">atoms</span> <span class="o">=</span> <span class="n">atoms</span>
            <span class="n">s1</span><span class="p">,</span> <span class="n">e1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span> <span class="n">e2</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">x</span><span class="o">.</span><span class="n">_vector</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="n">acetate</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span>
                                    <span class="n">acetate</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span>
                                    <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span>
                                    <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">]]]</span>
            <span class="n">translation</span><span class="p">,</span> <span class="n">angle</span><span class="p">,</span> <span class="n">axis</span><span class="p">,</span> <span class="n">point</span> <span class="o">=</span> <span class="n">find_transformations</span><span class="p">(</span>
                <span class="n">s1</span><span class="p">,</span> <span class="n">e1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span> <span class="n">e2</span><span class="p">,</span> <span class="n">radians</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">acetate</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="n">angle</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">,</span> <span class="n">point</span><span class="o">=</span><span class="n">point</span><span class="p">,</span> <span class="n">radians</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">acetate</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">vector</span><span class="o">=</span><span class="n">translation</span><span class="p">)</span>
            <span class="n">start_angle</span> <span class="o">=</span> <span class="n">dihedral</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">acetate</span><span class="p">[</span><span class="s1">&#39;O&#39;</span><span class="p">])</span>
            <span class="n">ref_angle</span> <span class="o">=</span> <span class="n">dihedral</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;O&#39;</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">cap_dihedral</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">False</span><span class="p">:</span>
                <span class="n">acetate</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span>
                    <span class="n">ref_angle</span> <span class="o">-</span> <span class="n">start_angle</span> <span class="o">+</span> <span class="n">cap_dihedral</span><span class="p">,</span>
                    <span class="n">axis</span><span class="o">=</span><span class="n">acetate</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span> <span class="o">-</span> <span class="n">acetate</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">,</span>
                    <span class="n">point</span><span class="o">=</span><span class="n">acetate</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">acetate</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span>
                    <span class="n">ref_angle</span> <span class="o">-</span> <span class="n">start_angle</span><span class="p">,</span>
                    <span class="n">axis</span><span class="o">=</span><span class="n">acetate</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span> <span class="o">-</span> <span class="n">acetate</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">,</span>
                    <span class="n">point</span><span class="o">=</span><span class="n">acetate</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">)</span>
            <span class="n">acetate</span><span class="p">[</span><span class="s1">&#39;OXT&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">ampal_parent</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;OXT&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">acetate</span><span class="p">[</span><span class="s1">&#39;OXT&#39;</span><span class="p">]</span>
            <span class="n">diff</span> <span class="o">=</span> <span class="n">acetate</span><span class="p">[</span><span class="s1">&#39;O&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;O&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;O&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span> <span class="o">+=</span> <span class="n">diff</span>
        <span class="k">elif</span> <span class="n">cap</span> <span class="o">==</span> <span class="s1">&#39;amide&#39;</span><span class="p">:</span>
            <span class="n">acetamide</span> <span class="o">=</span> <span class="n">Ligand</span><span class="p">(</span><span class="n">atoms</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mol_code</span><span class="o">=</span><span class="s1">&#39;UNK&#39;</span><span class="p">,</span> <span class="n">is_hetero</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">atoms</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Atom</span><span class="p">([</span><span class="o">-</span><span class="mf">0.4040</span><span class="p">,</span> <span class="mf">0.0000</span><span class="p">,</span> <span class="mf">1.4030</span><span class="p">],</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">res_label</span><span class="o">=</span><span class="s1">&#39;CA&#39;</span><span class="p">)</span>
            <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Atom</span><span class="p">([</span><span class="mf">0.0580</span><span class="p">,</span> <span class="mf">0.0000</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.0300</span><span class="p">],</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">res_label</span><span class="o">=</span><span class="s1">&#39;C&#39;</span><span class="p">)</span>
            <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;O&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Atom</span><span class="p">([</span><span class="mf">1.2440</span><span class="p">,</span> <span class="mf">0.0000</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.2840</span><span class="p">],</span> <span class="s1">&#39;O&#39;</span><span class="p">,</span> <span class="n">res_label</span><span class="o">=</span><span class="s1">&#39;O&#39;</span><span class="p">)</span>
            <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;NH2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Atom</span><span class="p">([</span><span class="o">-</span><span class="mf">0.8450</span><span class="p">,</span> <span class="mf">0.0000</span><span class="p">,</span> <span class="o">-</span><span class="mf">1.0300</span><span class="p">],</span>
                                <span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="n">res_label</span><span class="o">=</span><span class="s1">&#39;NH2&#39;</span><span class="p">)</span>
            <span class="n">acetamide</span><span class="o">.</span><span class="n">atoms</span> <span class="o">=</span> <span class="n">atoms</span>
            <span class="n">s1</span><span class="p">,</span> <span class="n">e1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span> <span class="n">e2</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">x</span><span class="o">.</span><span class="n">_vector</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="n">acetamide</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span>
                                    <span class="n">acetamide</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span>
                                    <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span>
                                    <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">]]]</span>
            <span class="n">translation</span><span class="p">,</span> <span class="n">angle</span><span class="p">,</span> <span class="n">axis</span><span class="p">,</span> <span class="n">point</span> <span class="o">=</span> <span class="n">find_transformations</span><span class="p">(</span>
                <span class="n">s1</span><span class="p">,</span> <span class="n">e1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span> <span class="n">e2</span><span class="p">,</span> <span class="n">radians</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">acetamide</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="n">angle</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">,</span>
                             <span class="n">point</span><span class="o">=</span><span class="n">point</span><span class="p">,</span> <span class="n">radians</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">acetamide</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">vector</span><span class="o">=</span><span class="n">translation</span><span class="p">)</span>
            <span class="n">start_angle</span> <span class="o">=</span> <span class="n">dihedral</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">acetamide</span><span class="p">[</span><span class="s1">&#39;O&#39;</span><span class="p">])</span>
            <span class="n">ref_angle</span> <span class="o">=</span> <span class="n">dihedral</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;O&#39;</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">cap_dihedral</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">False</span><span class="p">:</span>
                <span class="n">acetamide</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span>
                    <span class="n">ref_angle</span> <span class="o">-</span> <span class="n">start_angle</span> <span class="o">+</span> <span class="n">cap_dihedral</span><span class="p">,</span>
                    <span class="n">axis</span><span class="o">=</span><span class="n">acetamide</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span> <span class="o">-</span> <span class="n">acetamide</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">,</span>
                    <span class="n">point</span><span class="o">=</span><span class="n">acetamide</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">acetamide</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span>
                    <span class="n">ref_angle</span> <span class="o">-</span> <span class="n">start_angle</span><span class="p">,</span>
                    <span class="n">axis</span><span class="o">=</span><span class="n">acetamide</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span> <span class="o">-</span> <span class="n">acetamide</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">,</span>
                    <span class="n">point</span><span class="o">=</span><span class="n">acetamide</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ligands</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">ligands</span> <span class="o">=</span> <span class="n">LigandGroup</span><span class="p">(</span><span class="n">ampal_parent</span><span class="o">=</span><span class="bp">self</span><span class="p">)</span>
            <span class="n">amide</span> <span class="o">=</span> <span class="n">Ligand</span><span class="p">(</span><span class="n">mol_code</span><span class="o">=</span><span class="s1">&#39;NH2&#39;</span><span class="p">,</span> <span class="n">ampal_parent</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ligands</span><span class="p">)</span>
            <span class="n">amide_atoms</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">([(</span><span class="s1">&#39;NH2&#39;</span><span class="p">,</span> <span class="n">acetamide</span><span class="p">[</span><span class="s1">&#39;NH2&#39;</span><span class="p">])])</span>
            <span class="n">amide_atoms</span><span class="p">[</span><span class="s1">&#39;NH2&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">ampal_parent</span> <span class="o">=</span> <span class="n">amide</span>
            <span class="n">amide</span><span class="o">.</span><span class="n">atoms</span> <span class="o">=</span> <span class="n">amide_atoms</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ligands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">amide</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;assigned_ff&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">return</span></div>

<div class="viewcode-block" id="Polypeptide.n_cap"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Polypeptide.n_cap">[docs]</a>    <span class="k">def</span> <span class="nf">n_cap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n_cap</span><span class="o">=</span><span class="s1">&#39;acetyl&#39;</span><span class="p">,</span> <span class="n">cap_dihedral</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Adds an N-terminal acetamide cap.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Default behaviour is to duplicate the dihedral angle of the</span>
<span class="sd">        succeeding residues such that the orientation of the carbonyl</span>
<span class="sd">        of the acetyl will resemble that of the first residue. This</span>
<span class="sd">        can be adjusted by supplying a cap_dihedral value. Currently</span>
<span class="sd">        only acetyl cap is supported, but this structure should work</span>
<span class="sd">        for other caps.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        cap : str, optional</span>
<span class="sd">            Type of cap to be added. Options: &#39;acetyl&#39;</span>
<span class="sd">        cap_dihedral : bool</span>
<span class="sd">            Alternate psi angle to be used when added cap.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">n_cap</span> <span class="o">==</span> <span class="s1">&#39;acetyl&#39;</span><span class="p">:</span>
            <span class="n">methylacetamide</span> <span class="o">=</span> <span class="n">Ligand</span><span class="p">(</span>
                <span class="n">atoms</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mol_code</span><span class="o">=</span><span class="s1">&#39;UNK&#39;</span><span class="p">,</span> <span class="n">is_hetero</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">atoms</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Atom</span><span class="p">([</span><span class="mf">0.9500</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.2290</span><span class="p">,</span> <span class="mf">0.5090</span><span class="p">],</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">res_label</span><span class="o">=</span><span class="s1">&#39;C&#39;</span><span class="p">)</span>
            <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Atom</span><span class="p">([</span><span class="mf">0.7450</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.9430</span><span class="p">,</span> <span class="mf">1.8040</span><span class="p">],</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">res_label</span><span class="o">=</span><span class="s1">&#39;CA&#39;</span><span class="p">)</span>
            <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;O&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Atom</span><span class="p">([</span><span class="mf">0.1660</span><span class="p">,</span> <span class="o">-</span><span class="mf">2.0230</span><span class="p">,</span> <span class="mf">1.8130</span><span class="p">],</span> <span class="s1">&#39;O&#39;</span><span class="p">,</span> <span class="n">res_label</span><span class="o">=</span><span class="s1">&#39;O&#39;</span><span class="p">)</span>
            <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Atom</span><span class="p">([</span><span class="mf">1.2540</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.2750</span><span class="p">,</span> <span class="mf">2.9010</span><span class="p">],</span> <span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="n">res_label</span><span class="o">=</span><span class="s1">&#39;N&#39;</span><span class="p">)</span>
            <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;CME&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Atom</span><span class="p">([</span><span class="mf">1.1630</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.7870</span><span class="p">,</span> <span class="mf">4.2500</span><span class="p">],</span>
                                <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">res_label</span><span class="o">=</span><span class="s1">&#39;CME&#39;</span><span class="p">)</span>
            <span class="c1"># these coordinates seem ok, but could review</span>
            <span class="c1"># and use a different fragment if necessary</span>
            <span class="n">methylacetamide</span><span class="o">.</span><span class="n">atoms</span> <span class="o">=</span> <span class="n">atoms</span>
            <span class="n">s1</span><span class="p">,</span> <span class="n">e1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span> <span class="n">e2</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">x</span><span class="o">.</span><span class="n">_vector</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="n">methylacetamide</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">],</span>
                                    <span class="n">methylacetamide</span><span class="p">[</span><span class="s1">&#39;CME&#39;</span><span class="p">],</span>
                                    <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">],</span>
                                    <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">]]]</span>
            <span class="n">translation</span><span class="p">,</span> <span class="n">angle</span><span class="p">,</span> <span class="n">axis</span><span class="p">,</span> <span class="n">point</span> <span class="o">=</span> <span class="n">find_transformations</span><span class="p">(</span>
                <span class="n">s1</span><span class="p">,</span> <span class="n">e1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span> <span class="n">e2</span><span class="p">,</span> <span class="n">radians</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">methylacetamide</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span>
                <span class="n">angle</span><span class="o">=</span><span class="n">angle</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">,</span> <span class="n">point</span><span class="o">=</span><span class="n">point</span><span class="p">,</span> <span class="n">radians</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">methylacetamide</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">vector</span><span class="o">=</span><span class="n">translation</span><span class="p">)</span>
            <span class="n">start_angle</span> <span class="o">=</span> <span class="n">dihedral</span><span class="p">(</span>
                <span class="n">methylacetamide</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">])</span>
            <span class="n">ref_angle</span> <span class="o">=</span> <span class="n">dihedral</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;N&#39;</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;C&#39;</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">cap_dihedral</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">methylacetamide</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">ref_angle</span> <span class="o">-</span> <span class="n">start_angle</span> <span class="o">+</span> <span class="n">cap_dihedral</span><span class="p">,</span>
                                       <span class="n">axis</span><span class="o">=</span><span class="n">methylacetamide</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span> <span class="o">-</span>
                                       <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">,</span>
                                       <span class="n">point</span><span class="o">=</span><span class="n">methylacetamide</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">methylacetamide</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">ref_angle</span> <span class="o">-</span> <span class="n">start_angle</span><span class="p">,</span>
                                       <span class="n">axis</span><span class="o">=</span><span class="n">methylacetamide</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span> <span class="o">-</span>
                                       <span class="bp">self</span><span class="o">.</span><span class="n">_monomers</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">,</span>
                                       <span class="n">point</span><span class="o">=</span><span class="n">methylacetamide</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">_vector</span><span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ligands</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">ligands</span> <span class="o">=</span> <span class="n">LigandGroup</span><span class="p">(</span><span class="n">ampal_parent</span><span class="o">=</span><span class="bp">self</span><span class="p">)</span>
            <span class="n">acetamide</span> <span class="o">=</span> <span class="n">Ligand</span><span class="p">(</span><span class="n">mol_code</span><span class="o">=</span><span class="s1">&#39;ACM&#39;</span><span class="p">,</span> <span class="n">ampal_parent</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ligands</span><span class="p">)</span>
            <span class="n">acetamide_atoms</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="n">acetamide_atoms</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span>
            <span class="n">acetamide_atoms</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]</span>
            <span class="n">acetamide_atoms</span><span class="p">[</span><span class="s1">&#39;O&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;O&#39;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">acetamide_atoms</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="n">atom</span><span class="o">.</span><span class="n">ampal_parent</span> <span class="o">=</span> <span class="n">acetamide</span>
            <span class="n">acetamide</span><span class="o">.</span><span class="n">atoms</span> <span class="o">=</span> <span class="n">acetamide_atoms</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ligands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">acetamide</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span>  <span class="c1"># just in case we want to build different caps in later</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="s1">&#39;assigned_ff&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">return</span></div></div>


<div class="viewcode-block" id="Residue"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Residue">[docs]</a><span class="k">class</span> <span class="nc">Residue</span><span class="p">(</span><span class="n">Monomer</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Represents a amino acid `Residue`.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    atoms : OrderedDict, optional</span>
<span class="sd">        OrderedDict containing Atoms for the Monomer. OrderedDict</span>
<span class="sd">        is used to maintain the order items were added to the</span>
<span class="sd">        dictionary.</span>
<span class="sd">    mol_code : str, optional</span>
<span class="sd">        One or three letter code that represents the monomer.</span>
<span class="sd">    monomer_id : str, optional</span>
<span class="sd">        String used to identify the residue.</span>
<span class="sd">    insertion_code : str, optional</span>
<span class="sd">        Insertion code of monomer, used if reading from pdb.</span>
<span class="sd">    is_hetero : bool, optional</span>
<span class="sd">        True if is a hetero atom in pdb. Helps with PDB formatting.</span>
<span class="sd">    ampal_parent : ampal.Polypeptide, optional</span>
<span class="sd">        Reference to `Polypeptide` containing the `Residue`.</span>

<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    mol_code : str</span>
<span class="sd">        PDB molecule code that represents the `Residue`.</span>
<span class="sd">    insertion_code : str</span>
<span class="sd">        Insertion code of `Residue`, used if reading from pdb.</span>
<span class="sd">    is_hetero : bool</span>
<span class="sd">        True if is a hetero atom in pdb. Helps with PDB formatting.</span>
<span class="sd">    states : dict</span>
<span class="sd">        Contains an `OrderedDicts` containing atom information for each</span>
<span class="sd">        state available for the `Residue`.</span>
<span class="sd">    id : str</span>
<span class="sd">        String used to identify the residue.</span>
<span class="sd">    reference_atom : str</span>
<span class="sd">        The key that corresponds to the reference atom. This is used</span>
<span class="sd">        by various functions, for example backbone primitives are</span>
<span class="sd">        calculated using the atom defined using this key.</span>
<span class="sd">    ampal_parent : Polypeptide or None</span>
<span class="sd">        A reference to the `Polypeptide` containing this `Residue`.</span>
<span class="sd">    tags : dict</span>
<span class="sd">        A dictionary containing information about this AMPAL object.</span>
<span class="sd">        The tags dictionary is used by AMPAL to cache information</span>
<span class="sd">        about this object, but is also intended to be used by users</span>
<span class="sd">        to store any relevant information they have.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        Raised if `mol_code` is not length 1 or 3.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">atoms</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mol_code</span><span class="o">=</span><span class="s1">&#39;UNK&#39;</span><span class="p">,</span> <span class="n">monomer_id</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span>
                 <span class="n">insertion_code</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">is_hetero</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ampal_parent</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">Residue</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
            <span class="n">atoms</span><span class="p">,</span> <span class="n">monomer_id</span><span class="p">,</span> <span class="n">ampal_parent</span><span class="o">=</span><span class="n">ampal_parent</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mol_code</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mol_code</span> <span class="o">=</span> <span class="n">mol_code</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mol_letter</span> <span class="o">=</span> <span class="n">get_aa_letter</span><span class="p">(</span><span class="n">mol_code</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">mol_code</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mol_code</span> <span class="o">=</span> <span class="n">get_aa_code</span><span class="p">(</span><span class="n">mol_code</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mol_letter</span> <span class="o">=</span> <span class="n">mol_code</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s1">&#39;Monomer requires either a 1-letter or a 3-letter &#39;</span>
                <span class="s1">&#39;amino acid code (</span><span class="si">{}</span><span class="s1">)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mol_code</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">insertion_code</span> <span class="o">=</span> <span class="n">insertion_code</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_hetero</span> <span class="o">=</span> <span class="n">is_hetero</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference_atom</span> <span class="o">=</span> <span class="s1">&#39;CA&#39;</span>

    <span class="k">def</span> <span class="nf">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="s1">&#39;&lt;Residue containing </span><span class="si">{}</span><span class="s1"> </span><span class="si">{}</span><span class="s1">. Residue code: </span><span class="si">{}</span><span class="s1">&gt;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">atoms</span><span class="p">),</span> <span class="s1">&#39;Atom&#39;</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">atoms</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="k">else</span> <span class="s1">&#39;Atoms&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mol_code</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">backbone</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns a new `Residue` containing only the backbone atoms.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        bb_monomer : Residue</span>
<span class="sd">            `Residue` containing only the backbone atoms of the original</span>
<span class="sd">            `Monomer`.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        IndexError</span>
<span class="sd">            Raise if the `atoms` dict does not contain the backbone</span>
<span class="sd">            atoms (N, CA, C, O).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">backbone</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">([(</span><span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">]),</span>
                                    <span class="p">(</span><span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">]),</span>
                                    <span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]),</span>
                                    <span class="p">(</span><span class="s1">&#39;O&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;O&#39;</span><span class="p">])])</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">missing_atoms</span> <span class="o">=</span> <span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span>
                                   <span class="p">(</span><span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;O&#39;</span><span class="p">)</span>
                                   <span class="p">)</span>
            <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="s1">&#39;Error in residue </span><span class="si">{}</span><span class="s1"> </span><span class="si">{}</span><span class="s1"> </span><span class="si">{}</span><span class="s1">, missing (</span><span class="si">{}</span><span class="s1">) atoms. &#39;</span>
                           <span class="s1">&#39;`atoms` must be an `OrderedDict` with coordinates &#39;</span>
                           <span class="s1">&#39;defined for the backbone (N, CA, C, O) atoms.&#39;</span>
                           <span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ampal_parent</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mol_code</span><span class="p">,</span>
                                   <span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">missing_atoms</span><span class="p">)))</span>
        <span class="n">bb_monomer</span> <span class="o">=</span> <span class="n">Residue</span><span class="p">(</span><span class="n">backbone</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mol_code</span><span class="p">,</span> <span class="n">monomer_id</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                             <span class="n">insertion_code</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">insertion_code</span><span class="p">,</span>
                             <span class="n">is_hetero</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">is_hetero</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">bb_monomer</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">unique_id</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Generates a tuple that uniquely identifies a `Monomer` in an `Assembly`.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The unique_id will uniquely identify each monomer within a polymer.</span>
<span class="sd">        If each polymer in an assembly has a distinct id, it will uniquely</span>
<span class="sd">        identify each monomer within the assembly.</span>

<span class="sd">        The hetero-flag is defined as in Biopython as a string that is</span>
<span class="sd">        either a single whitespace in the case of a non-hetero atom,</span>
<span class="sd">        or &#39;H_&#39; plus the name of the hetero-residue (e.g. &#39;H_GLC&#39; in</span>
<span class="sd">        the case of a glucose molecule), or &#39;W&#39; in the case of a water</span>
<span class="sd">        molecule.</span>

<span class="sd">        For more information, see the Biopython documentation or this</span>
<span class="sd">        Biopython wiki page:</span>
<span class="sd">        http://biopython.org/wiki/The_Biopython_Structural_Bioinformatics_FAQ</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        unique_id : tuple</span>
<span class="sd">            unique_id[0] is the polymer_id unique_id[1] is a triple</span>
<span class="sd">            of the hetero-flag, the monomer id (residue number) and the</span>
<span class="sd">            insertion code.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_hetero</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mol_code</span> <span class="o">==</span> <span class="s1">&#39;HOH&#39;</span><span class="p">:</span>
                <span class="n">hetero_flag</span> <span class="o">=</span> <span class="s1">&#39;W&#39;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">hetero_flag</span> <span class="o">=</span> <span class="s1">&#39;H_</span><span class="si">{0}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mol_code</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">hetero_flag</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">ampal_parent</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="p">(</span><span class="n">hetero_flag</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">insertion_code</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">side_chain</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;List of the side-chain atoms (R-group).</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Returns empty list for glycine.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        side_chain_atoms: list(`Atoms`)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">side_chain_atoms</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mol_code</span> <span class="o">!=</span> <span class="s1">&#39;GLY&#39;</span><span class="p">:</span>
            <span class="n">covalent_bond_graph</span> <span class="o">=</span> <span class="n">generate_covalent_bond_graph</span><span class="p">(</span>
                <span class="n">find_covalent_bonds</span><span class="p">(</span><span class="bp">self</span><span class="p">))</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">subgraphs</span> <span class="o">=</span> <span class="n">generate_bond_subgraphs_from_break</span><span class="p">(</span>
                    <span class="n">covalent_bond_graph</span><span class="p">,</span> <span class="bp">self</span><span class="p">[</span><span class="s1">&#39;CA&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="p">[</span><span class="s1">&#39;CB&#39;</span><span class="p">])</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">subgraphs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">subgraphs</span> <span class="o">=</span> <span class="n">generate_bond_subgraphs_from_break</span><span class="p">(</span>
                        <span class="n">subgraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="bp">self</span><span class="p">[</span><span class="s1">&#39;CD&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">])</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">subgraphs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">subgraphs</span><span class="p">:</span>
                        <span class="k">if</span> <span class="bp">self</span><span class="p">[</span><span class="s1">&#39;CB&#39;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">g</span><span class="p">:</span>
                            <span class="n">side_chain_atoms</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">nodes</span><span class="p">()</span>
                            <span class="k">break</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="n">warning_message</span> <span class="o">=</span> <span class="s2">&quot;Malformed PDB for Residue </span><span class="si">{0}</span><span class="s2">: </span><span class="si">{1}</span><span class="s2">.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span>
                <span class="k">if</span> <span class="s1">&#39;CB&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="n">side_chain_atoms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="p">[</span><span class="s1">&#39;CB&#39;</span><span class="p">])</span>
                    <span class="n">warning_message</span> <span class="o">+=</span> <span class="s2">&quot; Side-chain is just the CB atom.&quot;</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">warning_message</span> <span class="o">+=</span> <span class="s2">&quot; Empty side-chain.&quot;</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="n">warning_message</span><span class="p">,</span> <span class="n">MalformedPDBWarning</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">side_chain_atoms</span>

    <span class="c1"># TODO fix behaviour to allow option not to include residue itself</span>
<div class="viewcode-block" id="Residue.side_chain_environment"><a class="viewcode-back" href="../../../isambard.ampal.protein.html#isambard.ampal.protein.Residue.side_chain_environment">[docs]</a>    <span class="k">def</span> <span class="nf">side_chain_environment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cutoff</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">include_neighbours</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                               <span class="n">inter_chain</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">include_ligands</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">include_solvent</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Finds `Residues` with any atom within the cutoff distance of side-chain.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Includes the parent residue in the list.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        cutoff : float, optional</span>
<span class="sd">            Maximum inter-atom distance for residue to be included.</span>
<span class="sd">            Defaults to 4.</span>
<span class="sd">        include_neighbours : bool, optional</span>
<span class="sd">            If `false`, does not return `Residue` at i-1, i+1 positions</span>
<span class="sd">            in same chain as `Residue`.</span>
<span class="sd">        inter_chain : bool, optional</span>
<span class="sd">            If `false`, only includes nearby `Residue` in the same chain</span>
<span class="sd">            as the `Residue`.</span>
<span class="sd">        include_ligands : bool, optional</span>
<span class="sd">            If `true`, `Residue` classed as ligands but not identified as</span>
<span class="sd">            solvent will be included in the environment.</span>
<span class="sd">        include_solvent : bool, optional</span>
<span class="sd">            If `true`, Monomers classed as categorised as solvent</span>
<span class="sd">            will be included in the environment.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        sc_environment : list</span>
<span class="sd">            List of monomers within cutoff distance of side-chain.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mol_code</span> <span class="o">==</span> <span class="s1">&#39;GLY&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">[</span><span class="bp">self</span><span class="p">]</span>
        <span class="n">side_chain_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">:</span> <span class="p">{</span><span class="n">y</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">states</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="n">y</span><span class="p">]</span>
                               <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">states</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">states</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="n">y</span><span class="p">]</span> <span class="ow">in</span>
                               <span class="bp">self</span><span class="o">.</span><span class="n">side_chain</span><span class="p">}</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">states</span><span class="p">}</span>
        <span class="n">side_chain_monomer</span> <span class="o">=</span> <span class="n">Monomer</span><span class="p">(</span>
            <span class="n">atoms</span><span class="o">=</span><span class="n">side_chain_dict</span><span class="p">,</span> <span class="n">monomer_id</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
            <span class="n">ampal_parent</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ampal_parent</span><span class="p">)</span>
        <span class="n">sc_environment</span> <span class="o">=</span> <span class="n">side_chain_monomer</span><span class="o">.</span><span class="n">environment</span><span class="p">(</span>
            <span class="n">cutoff</span><span class="o">=</span><span class="n">cutoff</span><span class="p">,</span> <span class="n">include_ligands</span><span class="o">=</span><span class="n">include_ligands</span><span class="p">,</span>
            <span class="n">include_neighbours</span><span class="o">=</span><span class="n">include_neighbours</span><span class="p">,</span>
            <span class="n">include_solvent</span><span class="o">=</span><span class="n">include_solvent</span><span class="p">,</span> <span class="n">inter_chain</span><span class="o">=</span><span class="n">inter_chain</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">sc_environment</span></div></div>


<span class="n">__author__</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;Jack W. Heal, Christopher W. Wood, Gail J. Bartlett, &#39;</span>
              <span class="s1">&#39;Andrew R. Thomson, Kieran L. Hudson&#39;</span><span class="p">)</span>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Woolfson Group.
      Last updated on October 09, 2017.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'1.4.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>