

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>isambard.ampal.protein module &mdash; ISAMBARD 1.4.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="ISAMBARD 1.4.1 documentation" href="index.html"/>
        <link rel="up" title="isambard.ampal package" href="isambard.ampal.html"/>
        <link rel="next" title="isambard.ampal.pseudo_atoms module" href="isambard.ampal.pseudo_atoms.html"/>
        <link rel="prev" title="isambard.ampal.pdb_parser module" href="isambard.ampal.pdb_parser.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> ISAMBARD
          

          
          </a>

          
            
            
              <div class="version">
                2016.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_reference.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="developer_guide.html">Developer Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="citing_isambard.html">Citing ISAMBARD</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ISAMBARD</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="api_reference.html">API Reference</a> &raquo;</li>
        
          <li><a href="isambard.ampal.html">isambard.ampal package</a> &raquo;</li>
        
      <li>isambard.ampal.protein module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/isambard.ampal.protein.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-isambard.ampal.protein">
<span id="isambard-ampal-protein-module"></span><h1>isambard.ampal.protein module<a class="headerlink" href="#module-isambard.ampal.protein" title="Permalink to this headline">¶</a></h1>
<p>AMPAL objects that represent protein.</p>
<dl class="class">
<dt id="isambard.ampal.protein.Polypeptide">
<em class="property">class </em><code class="descname">Polypeptide</code><span class="sig-paren">(</span><em>monomers=None</em>, <em>polymer_id=' '</em>, <em>ampal_parent=None</em>, <em>sl=2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">ampal.base_ampal.Polymer</span></code></p>
<p>Container for <cite>Residues</cite>, inherits from <cite>Polymer</cite>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>monomers</strong> (<a class="reference internal" href="#isambard.ampal.protein.Residue" title="isambard.ampal.protein.Residue"><em>Residue</em></a><em> or </em><em>[</em><a class="reference internal" href="#isambard.ampal.protein.Residue" title="isambard.ampal.protein.Residue"><em>Residue</em></a><em>]</em><em>, </em><em>optional</em>) – <cite>Residue</cite> or list containing <cite>Residue</cite> objects to form the
<cite>Polypeptide</cite>.</li>
<li><strong>polymer_id</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a><em>, </em><em>optional</em>) – An ID that the user can use to identify the <cite>Polypeptide</cite>. This is
used when generating a pdb file using <cite>Polypeptide().pdb</cite>.</li>
<li><strong>ampal_parent</strong> (<em>ampal.Assembly</em><em>, </em><em>optional</em>) – Reference to <cite>Assembly</cite> containing the <cite>Polymer</cite>.</li>
<li><strong>sl</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a><em>, </em><em>optional</em>) – The default smoothing level used when calculating the
backbone primitive.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.id">
<code class="descname">id</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.id" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> – <cite>Polypeptide</cite> ID</p>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.ampal_parent">
<code class="descname">ampal_parent</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.ampal_parent" title="Permalink to this definition">¶</a></dt>
<dd><p><em>ampal.Assembly or None</em> – Reference to <cite>Assembly</cite> containing the <cite>Polypeptide</cite></p>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.molecule_type">
<code class="descname">molecule_type</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.molecule_type" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> – A description of the type of <cite>Polymer</cite> i.e. Protein, DNA etc.</p>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.ligands">
<code class="descname">ligands</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.ligands" title="Permalink to this definition">¶</a></dt>
<dd><p><em>ampal.LigandGroup</em> – A <cite>LigandGroup</cite> containing all the <cite>Ligands</cite> associated with this
<cite>Polypeptide</cite> chain.</p>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.tags">
<code class="descname">tags</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.tags" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dict</em> – A dictionary containing information about this AMPAL object.
The tags dictionary is used by AMPAL to cache information
about this object, but is also intended to be used by users
to store any relevant information they have.</p>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.sl">
<code class="descname">sl</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.sl" title="Permalink to this definition">¶</a></dt>
<dd><p><em>int</em> – The default smoothing level used when calculating the
backbone primitive.</p>
</dd></dl>

<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">TypeError</span></code> – <cite>Polymer</cite> type objects can only be initialised empty or using
a <cite>Monomer</cite>.</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.backbone">
<code class="descname">backbone</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.backbone" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a new <cite>Polymer</cite> containing only the backbone atoms.</p>
<p class="rubric">Notes</p>
<p>Metadata is not currently preserved from the parent object.
Sequence data is retained, but only the main chain atoms are retained.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><strong>bb_poly</strong> – Polymer containing only the backbone atoms of the original
Polymer.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#isambard.ampal.protein.Polypeptide" title="isambard.ampal.protein.Polypeptide">Polypeptide</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.backbone_bond_angles">
<code class="descname">backbone_bond_angles</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.backbone_bond_angles" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary containing backbone bond angles as lists of floats.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><strong>bond_angles</strong> – Keys are <cite>n_ca_c</cite>, <cite>ca_c_o</cite>, <cite>ca_c_n</cite> and <cite>c_n_ca</cite>, referring
to the N-CA-C, CA-C=O, CA-C-N and C-N-CA angles respectively.
Values are lists of floats : the bond angles in degrees.
The lists of n_ca_c, ca_c_o are of length k for a <cite>Polypeptide</cite>
containing k <cite>Residues</cite>. The list of ca_c_n and c_n_ca are of
length k-1 for a <cite>Polypeptide</cite> containing k <cite>Residues</cite> (These
angles are across the peptide bond, and are therefore formed
between successive <cite>Residue</cite> pairs).</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/2/library/stdtypes.html#dict" title="(in Python v2.7)">dict</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.backbone_bond_lengths">
<code class="descname">backbone_bond_lengths</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.backbone_bond_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary containing backbone bond lengths as lists of floats.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><strong>bond_lengths</strong> – Keys are <cite>n_ca</cite>, <cite>ca_c</cite>, <cite>c_o</cite> and <cite>c_n</cite>, referring to the
N-CA, CA-C, C=O and C-N bonds respectively. Values are
lists of floats : the bond lengths in Angstroms.
The lists of n_ca, ca_c and c_o are of length k for
a Polypeptide containing k Residues. The list of c_n bonds
is of length k-1 for a Polypeptide containing k Residues
(C-N formed between successive <cite>Residue</cite> pairs).</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/2/library/stdtypes.html#dict" title="(in Python v2.7)">dict</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.c_cap">
<code class="descname">c_cap</code><span class="sig-paren">(</span><em>cap='acid'</em>, <em>cap_dihedral=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.c_cap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.c_cap" title="Permalink to this definition">¶</a></dt>
<dd><p>Caps C-terminus of polypeptide chain.</p>
<p class="rubric">Notes</p>
<p>Default behaviour is to add an oxygen atom to create a
carboxylate function at the C-terminus without changing the
psi angle of the C-terminal residue. Alternative psi angles
can be accessed through the cap_dihedral parameter. Will not
remove an existing cap if one is present, though altering a
cap of the same type will overwrite the original one.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>cap</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a><em>, </em><em>optional</em>) – Type of cap to be added. Options: ‘acid’, ‘amide’</li>
<li><strong>cap_dihedral</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a>) – Alternate psi angle to be used when added cap.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.c_join">
<code class="descname">c_join</code><span class="sig-paren">(</span><em>other</em>, <em>psi=-40.76</em>, <em>omega=-178.25</em>, <em>phi=-65.07</em>, <em>o_c_n_angle=None</em>, <em>c_n_ca_angle=None</em>, <em>c_n_length=None</em>, <em>relabel=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.c_join"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.c_join" title="Permalink to this definition">¶</a></dt>
<dd><p>Joins other to self at the C-terminus via a peptide bond.</p>
<p class="rubric">Notes</p>
<p>This function directly modifies self. It does not return a new object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>other</strong> (<a class="reference internal" href="#isambard.ampal.protein.Residue" title="isambard.ampal.protein.Residue"><em>Residue</em></a><em> or </em><a class="reference internal" href="#isambard.ampal.protein.Polypeptide" title="isambard.ampal.protein.Polypeptide"><em>Polypeptide</em></a>) – </li>
<li><strong>psi</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a><em>, </em><em>optional</em>) – Psi torsion angle (degrees) between final <cite>Residue</cite> of self
and first <cite>Residue</cite> of other.</li>
<li><strong>omega</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a><em>, </em><em>optional</em>) – Omega torsion angle (degrees) between final <cite>Residue</cite> of
self and first <cite>Residue</cite> of other.</li>
<li><strong>phi</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a><em>, </em><em>optional</em>) – Phi torsion angle (degrees) between final <cite>Residue</cite> of self
and first <cite>Residue</cite> of other.</li>
<li><strong>o_c_n_angle</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a><em> or </em><a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a><em>, </em><em>optional</em>) – Desired angle between O, C (final <cite>Residue</cite> of self) and N
(first <cite>Residue</cite> of other) atoms. If <cite>None</cite>, default value is
taken from <cite>ideal_backbone_bond_angles</cite>.</li>
<li><strong>c_n_ca_angle</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a><em> or </em><a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a><em>, </em><em>optional</em>) – Desired angle between C (final <cite>Residue</cite> of self) and N, CA
(first <cite>Residue</cite> of other) atoms. If <cite>None</cite>, default value is
taken from <cite>ideal_backbone_bond_angles</cite>.</li>
<li><strong>c_n_length</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a><em> or </em><a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a><em>, </em><em>optional</em>) – Desired peptide bond length between final <cite>Residue</cite> of self
and first <cite>Residue</cite> of other. If <cite>None</cite>, default value is taken
from <cite>ideal_backbone_bond_lengths</cite>.</li>
<li><strong>relabel</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a><em>, </em><em>optional</em>) – If <cite>True</cite>, <cite>relabel_all</cite> is run on self before returning.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last">TypeError: – If other is not a <cite>Residue</cite> or a Polypeptide.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.fasta">
<code class="descname">fasta</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates sequence data for the protein in FASTA format.</p>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.get_slice_from_res_id">
<code class="descname">get_slice_from_res_id</code><span class="sig-paren">(</span><em>start</em>, <em>end</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.get_slice_from_res_id"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.get_slice_from_res_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a new <cite>Polypeptide</cite> containing the <cite>Residues</cite> in start/end range.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>start</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) – string representing start residue id (PDB numbering)</li>
<li><strong>end</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) – string representing end residue id (PDB numbering)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>slice_polymer</strong> – Polymer containing the residue range specified by start-end</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="isambard.ampal.base_ampal.html#isambard.ampal.base_ampal.Polymer" title="isambard.ampal.base_ampal.Polymer">Polymer</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.helices">
<code class="descname">helices</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.helices" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a new <cite>Assembly</cite> containing only the alpha-helices.</p>
<p class="rubric">Notes</p>
<p>Metadata is not currently preserved from the parent object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><strong>hel_assembly</strong> – <cite>Assembly</cite> containing only the alpha-helices of the
original <cite>Polymer</cite>.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="isambard.ampal.assembly.html#isambard.ampal.assembly.Assembly" title="isambard.ampal.assembly.Assembly">Assembly</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.isoelectric_point">
<code class="descname">isoelectric_point</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.isoelectric_point" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the isoelectric point of the <cite>Assembly</cite>.</p>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.molar_extinction_280">
<code class="descname">molar_extinction_280</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.molar_extinction_280" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the extinction co-efficient of the <cite>Assembly</cite> at 280 nm.</p>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.molecular_weight">
<code class="descname">molecular_weight</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.molecular_weight" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the molecular weight of the <cite>Assembly</cite> in Daltons.</p>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.n_cap">
<code class="descname">n_cap</code><span class="sig-paren">(</span><em>n_cap='acetyl'</em>, <em>cap_dihedral=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.n_cap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.n_cap" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds an N-terminal acetamide cap.</p>
<p class="rubric">Notes</p>
<p>Default behaviour is to duplicate the dihedral angle of the
succeeding residues such that the orientation of the carbonyl
of the acetyl will resemble that of the first residue. This
can be adjusted by supplying a cap_dihedral value. Currently
only acetyl cap is supported, but this structure should work
for other caps.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>cap</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a><em>, </em><em>optional</em>) – Type of cap to be added. Options: ‘acetyl’</li>
<li><strong>cap_dihedral</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a>) – Alternate psi angle to be used when added cap.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.n_join">
<code class="descname">n_join</code><span class="sig-paren">(</span><em>other</em>, <em>psi=-40.76</em>, <em>omega=-178.25</em>, <em>phi=-65.07</em>, <em>o_c_n_angle=None</em>, <em>c_n_ca_angle=None</em>, <em>c_n_length=None</em>, <em>relabel=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.n_join"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.n_join" title="Permalink to this definition">¶</a></dt>
<dd><p>Joins other to self at the N-terminus via a peptide bond.</p>
<p class="rubric">Notes</p>
<p>This function directly modifies self. It does not return a new object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>other</strong> (<a class="reference internal" href="#isambard.ampal.protein.Residue" title="isambard.ampal.protein.Residue"><em>Residue</em></a><em> or </em><a class="reference internal" href="#isambard.ampal.protein.Polypeptide" title="isambard.ampal.protein.Polypeptide"><em>Polypeptide</em></a>) – </li>
<li><strong>psi</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) – Psi torsion angle (degrees) between final <cite>Residue</cite> of other
and first <cite>Residue</cite> of self.</li>
<li><strong>omega</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) – Omega torsion angle (degrees) between final <cite>Residue</cite> of
other and first <cite>Residue</cite> of self.</li>
<li><strong>phi</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) – Phi torsion angle (degrees) between final <cite>Residue</cite> of other
and first <cite>Residue</cite> of self.</li>
<li><strong>o_c_n_angle</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a><em> or </em><a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) – Desired angle between O, C (final <cite>Residue</cite> of other) and N
(first <cite>Residue</cite> of self) atoms. If <cite>None</cite>, default value is
taken from <cite>ideal_backbone_bond_angles</cite>.</li>
<li><strong>c_n_ca_angle</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a><em> or </em><a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) – Desired angle between C (final <cite>Residue</cite> of other) and N, CA
(first <cite>Residue</cite> of self) atoms. If <cite>None</cite>, default value is taken
from <cite>ideal_backbone_bond_angles</cite>.</li>
<li><strong>c_n_length</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a><em> or </em><a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) – Desired peptide bond length between final <cite>Residue</cite> of other
and first <cite>Residue</cite> of self. If None, default value is taken
from ideal_backbone_bond_lengths.</li>
<li><strong>relabel</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a>) – If True, relabel_all is run on self before returning.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last">TypeError: – If other is not a <cite>Residue</cite> or a <cite>Polypeptide</cite></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.pack_new_sequence">
<code class="descname">pack_new_sequence</code><span class="sig-paren">(</span><em>sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.pack_new_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.pack_new_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Packs a new sequence onto the polymer using Scwrl4.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>sequence</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) – String containing the amino acid sequence. This must
be the same length as the Polymer</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> – Raised if the sequence length does not match the
number of monomers in the Polymer.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.primitive">
<code class="descname">primitive</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.primitive" title="Permalink to this definition">¶</a></dt>
<dd><p>Primitive of the backbone.</p>
<p class="rubric">Notes</p>
<p>This is the average of the positions of all the CAs in frames
of <cite>sl</cite> <cite>Residues</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.radii_of_curvature">
<code class="descname">radii_of_curvature</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.radii_of_curvature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.radii_of_curvature" title="Permalink to this definition">¶</a></dt>
<dd><p>List of radius of curvature values along the <cite>Polypeptide</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.repack">
<code class="descname">repack</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.repack"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.repack" title="Permalink to this definition">¶</a></dt>
<dd><p>Repacks current side chain sequence using Scwrl4.</p>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.rise_per_residue">
<code class="descname">rise_per_residue</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.rise_per_residue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.rise_per_residue" title="Permalink to this definition">¶</a></dt>
<dd><p>List of rise per residue values along the <cite>Polypeptide</cite>.
.. rubric:: Notes</p>
<p>Calculated from <cite>Polypeptide.primitive</cite>.</p>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.sequence">
<code class="descname">sequence</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the sequence of the <cite>Polymer</cite> as a string.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><strong>sequence</strong> – String of the <cite>Residue</cite> sequence of the <cite>Polypeptide</cite>.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)">str</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Polypeptide.strands">
<code class="descname">strands</code><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.strands" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a new <cite>Assembly</cite> containing only the beta-strand atoms.</p>
<p class="rubric">Notes</p>
<p>Metadata is not currently preserved from the parent object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><strong>strand_assembly</strong> – <cite>Assembly</cite> containing only the beta-strand atoms of
the original <cite>Polymer</cite>.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="isambard.ampal.assembly.html#isambard.ampal.assembly.Assembly" title="isambard.ampal.assembly.Assembly">Assembly</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.tag_ca_geometry">
<code class="descname">tag_ca_geometry</code><span class="sig-paren">(</span><em>force=False</em>, <em>reference_axis=None</em>, <em>reference_axis_name='ref_axis'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.tag_ca_geometry"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.tag_ca_geometry" title="Permalink to this definition">¶</a></dt>
<dd><p>Tags each <cite>Residue</cite> with rise_per_residue, radius_of_curvature and residues_per_turn.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>force</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a><em>, </em><em>optional</em>) – If <cite>True</cite> the tag will be run even if <cite>Residues</cite> are already
tagged.</li>
<li><strong>reference_axis</strong> (<em>list</em><em>(</em><em>numpy.array</em><em> or </em><a class="reference external" href="https://docs.python.org/2/library/functions.html#tuple" title="(in Python v2.7)"><em>tuple</em></a><em> or </em><em>list</em><em>)</em><em>, </em><em>optional</em>) – Coordinates to feed to geometry functions that depend on
having a reference axis.</li>
<li><strong>reference_axis_name</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a><em>, </em><em>optional</em>) – Used to name the keys in tags at <cite>Polypeptide</cite> and <cite>Residue</cite> level.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.tag_dssp_solvent_accessibility">
<code class="descname">tag_dssp_solvent_accessibility</code><span class="sig-paren">(</span><em>force=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.tag_dssp_solvent_accessibility"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.tag_dssp_solvent_accessibility" title="Permalink to this definition">¶</a></dt>
<dd><p>Tags each <cite>Residues</cite> Polymer with its solvent accessibility.</p>
<p class="rubric">Notes</p>
<dl class="docutils">
<dt>For more about DSSP’s solvent accessibilty metric, see:</dt>
<dd><a class="reference external" href="http://swift.cmbi.ru.nl/gv/dssp/HTML/descrip.html#ACC">http://swift.cmbi.ru.nl/gv/dssp/HTML/descrip.html#ACC</a></dd>
</dl>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id1" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[1]</td><td>Kabsch W, Sander C (1983) “Dictionary of protein
secondary structure: pattern recognition of hydrogen-bonded
and geometrical features”, Biopolymers, 22, 2577-637.</td></tr>
</tbody>
</table>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>force</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a><em>, </em><em>optional</em>) – If <cite>True</cite> the tag will be run even if <cite>Residues</cite> are
already tagged.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.tag_residue_solvent_accessibility">
<code class="descname">tag_residue_solvent_accessibility</code><span class="sig-paren">(</span><em>tag_type=False</em>, <em>tag_total=False</em>, <em>force=False</em>, <em>include_hetatms=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.tag_residue_solvent_accessibility"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.tag_residue_solvent_accessibility" title="Permalink to this definition">¶</a></dt>
<dd><p>Tags <cite>Residues</cite> wirh relative residue solvent accessibility.</p>
<p class="rubric">Notes</p>
<p>THIS FUNCTIONALITY REQUIRES NACESS.
This function tags the Monomer with the <em>relative</em> RSA of
the <em>whole side chain</em>, i.e. column 2 of the .rsa file that
NACCESS writes.</p>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id2" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[1]</td><td>Hubbard,S.J. &amp; Thornton, J.M. (1993), ‘NACCESS’,
Computer Program, Department of Biochemistry and Molecular
Biology, University College London.</td></tr>
</tbody>
</table>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>force</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a><em>, </em><em>optional</em>) – If <cite>True</cite>, the ta will be run even if <cite>Residues</cite> are
already tagged.</li>
<li><strong>tag_type</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a><em>, </em><em>optional</em>) – Specifies the name of the tag. Defaults to
‘residue_solvent_accessibility’. Useful for specifying more
than one tag, e.g. if the Polymer is part of an Assembly.</li>
<li><strong>tag_total</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a><em>, </em><em>optional</em>) – If True then the total rsa of the Polymer will be tagged
in the ‘total accessibility’ tag.</li>
<li><strong>include_hetatms</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a><em>, </em><em>optional</em>) – If true then NACCESS will run with the -h flag and will
include heteroatom solvent accessibility where it can.
Helpful if your file has MSE residues that you don’t
convert to MET, but best check if they are there
before using the flag.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.tag_secondary_structure">
<code class="descname">tag_secondary_structure</code><span class="sig-paren">(</span><em>force=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.tag_secondary_structure"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.tag_secondary_structure" title="Permalink to this definition">¶</a></dt>
<dd><p>Tags each <cite>Residue</cite> of the <cite>Polypeptide</cite> with secondary structure.</p>
<p class="rubric">Notes</p>
<p>DSSP must be available to call. Check by running
<cite>isambard.external_programs.dssp.test_dssp</cite>. If DSSP is not
available, please follow instruction here to add it:
<a class="reference external" href="https://github.com/woolfson-group/isambard#external-programs">https://github.com/woolfson-group/isambard#external-programs</a></p>
<p>For more information on DSSP see [1].</p>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id3" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[1]</td><td>Kabsch W, Sander C (1983) “Dictionary of protein
secondary structure: pattern recognition of hydrogen-bonded
and geometrical features”, Biopolymers, 22, 2577-637.</td></tr>
</tbody>
</table>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>force</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a><em>, </em><em>optional</em>) – If <cite>True</cite> the tag will be run even if <cite>Residues</cite> are
already tagged.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.tag_sidechain_dihedrals">
<code class="descname">tag_sidechain_dihedrals</code><span class="sig-paren">(</span><em>force=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.tag_sidechain_dihedrals"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.tag_sidechain_dihedrals" title="Permalink to this definition">¶</a></dt>
<dd><p>Tags each monomer with side-chain dihedral angles</p>
<dl class="docutils">
<dt>force: bool, optional</dt>
<dd>If <cite>True</cite> the tag will be run even if <cite>Residues</cite> are
already tagged.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.tag_torsion_angles">
<code class="descname">tag_torsion_angles</code><span class="sig-paren">(</span><em>force=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.tag_torsion_angles"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.tag_torsion_angles" title="Permalink to this definition">¶</a></dt>
<dd><p>Tags each Monomer of the Polymer with its omega, phi and psi torsion angle.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>force</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a><em>, </em><em>optional</em>) – If <cite>True</cite> the tag will be run even if <cite>Residues</cite> are
already tagged.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.valid_backbone_bond_angles">
<code class="descname">valid_backbone_bond_angles</code><span class="sig-paren">(</span><em>atol=20</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.valid_backbone_bond_angles"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.valid_backbone_bond_angles" title="Permalink to this definition">¶</a></dt>
<dd><p>True if all backbone bond angles are within atol degrees of their expected values.</p>
<p class="rubric">Notes</p>
<p>Ideal bond angles taken from [1].</p>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id4" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[1]</td><td>Schulz, G. E, and R. Heiner Schirmer. Principles Of
Protein Structure. New York: Springer-Verlag, 1979.</td></tr>
</tbody>
</table>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>atol</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a><em>, </em><em>optional</em>) – Tolerance value in degrees for the absolute deviation
away from ideal backbone bond angles.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Polypeptide.valid_backbone_bond_lengths">
<code class="descname">valid_backbone_bond_lengths</code><span class="sig-paren">(</span><em>atol=0.1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Polypeptide.valid_backbone_bond_lengths"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Polypeptide.valid_backbone_bond_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>True if all backbone bonds are within atol Angstroms of the expected distance.</p>
<p class="rubric">Notes</p>
<p>Ideal bond lengths taken from [1].</p>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id5" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[1]</td><td>Schulz, G. E, and R. Heiner Schirmer. Principles Of
Protein Structure. New York: Springer-Verlag, 1979.</td></tr>
</tbody>
</table>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>atol</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a><em>, </em><em>optional</em>) – Tolerance value in Angstoms for the absolute deviation
away from ideal backbone bond lengths.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="isambard.ampal.protein.Residue">
<em class="property">class </em><code class="descname">Residue</code><span class="sig-paren">(</span><em>atoms=None</em>, <em>mol_code='UNK'</em>, <em>monomer_id=' '</em>, <em>insertion_code=' '</em>, <em>is_hetero=False</em>, <em>ampal_parent=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Residue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Residue" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">ampal.base_ampal.Monomer</span></code></p>
<p>Represents a amino acid <cite>Residue</cite>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>atoms</strong> (<em>OrderedDict</em><em>, </em><em>optional</em>) – OrderedDict containing Atoms for the Monomer. OrderedDict
is used to maintain the order items were added to the
dictionary.</li>
<li><strong>mol_code</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a><em>, </em><em>optional</em>) – One or three letter code that represents the monomer.</li>
<li><strong>monomer_id</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a><em>, </em><em>optional</em>) – String used to identify the residue.</li>
<li><strong>insertion_code</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a><em>, </em><em>optional</em>) – Insertion code of monomer, used if reading from pdb.</li>
<li><strong>is_hetero</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a><em>, </em><em>optional</em>) – True if is a hetero atom in pdb. Helps with PDB formatting.</li>
<li><strong>ampal_parent</strong> (<em>ampal.Polypeptide</em><em>, </em><em>optional</em>) – Reference to <cite>Polypeptide</cite> containing the <cite>Residue</cite>.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="isambard.ampal.protein.Residue.mol_code">
<code class="descname">mol_code</code><a class="headerlink" href="#isambard.ampal.protein.Residue.mol_code" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> – PDB molecule code that represents the <cite>Residue</cite>.</p>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Residue.insertion_code">
<code class="descname">insertion_code</code><a class="headerlink" href="#isambard.ampal.protein.Residue.insertion_code" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> – Insertion code of <cite>Residue</cite>, used if reading from pdb.</p>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Residue.is_hetero">
<code class="descname">is_hetero</code><a class="headerlink" href="#isambard.ampal.protein.Residue.is_hetero" title="Permalink to this definition">¶</a></dt>
<dd><p><em>bool</em> – True if is a hetero atom in pdb. Helps with PDB formatting.</p>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Residue.states">
<code class="descname">states</code><a class="headerlink" href="#isambard.ampal.protein.Residue.states" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dict</em> – Contains an <cite>OrderedDicts</cite> containing atom information for each
state available for the <cite>Residue</cite>.</p>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Residue.id">
<code class="descname">id</code><a class="headerlink" href="#isambard.ampal.protein.Residue.id" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> – String used to identify the residue.</p>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Residue.reference_atom">
<code class="descname">reference_atom</code><a class="headerlink" href="#isambard.ampal.protein.Residue.reference_atom" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> – The key that corresponds to the reference atom. This is used
by various functions, for example backbone primitives are
calculated using the atom defined using this key.</p>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Residue.ampal_parent">
<code class="descname">ampal_parent</code><a class="headerlink" href="#isambard.ampal.protein.Residue.ampal_parent" title="Permalink to this definition">¶</a></dt>
<dd><p><em>Polypeptide or None</em> – A reference to the <cite>Polypeptide</cite> containing this <cite>Residue</cite>.</p>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Residue.tags">
<code class="descname">tags</code><a class="headerlink" href="#isambard.ampal.protein.Residue.tags" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dict</em> – A dictionary containing information about this AMPAL object.
The tags dictionary is used by AMPAL to cache information
about this object, but is also intended to be used by users
to store any relevant information they have.</p>
</dd></dl>

<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> – Raised if <cite>mol_code</cite> is not length 1 or 3.</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="isambard.ampal.protein.Residue.backbone">
<code class="descname">backbone</code><a class="headerlink" href="#isambard.ampal.protein.Residue.backbone" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a new <cite>Residue</cite> containing only the backbone atoms.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><strong>bb_monomer</strong> – <cite>Residue</cite> containing only the backbone atoms of the original
<cite>Monomer</cite>.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#isambard.ampal.protein.Residue" title="isambard.ampal.protein.Residue">Residue</a></td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">IndexError</span></code> – Raise if the <cite>atoms</cite> dict does not contain the backbone
atoms (N, CA, C, O).</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Residue.side_chain">
<code class="descname">side_chain</code><a class="headerlink" href="#isambard.ampal.protein.Residue.side_chain" title="Permalink to this definition">¶</a></dt>
<dd><p>List of the side-chain atoms (R-group).</p>
<p class="rubric">Notes</p>
<p>Returns empty list for glycine.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><strong>side_chain_atoms</strong></td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list(<cite>Atoms</cite>)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="isambard.ampal.protein.Residue.side_chain_environment">
<code class="descname">side_chain_environment</code><span class="sig-paren">(</span><em>cutoff=4</em>, <em>include_neighbours=True</em>, <em>inter_chain=True</em>, <em>include_ligands=False</em>, <em>include_solvent=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#Residue.side_chain_environment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.Residue.side_chain_environment" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds <cite>Residues</cite> with any atom within the cutoff distance of side-chain.</p>
<p class="rubric">Notes</p>
<p>Includes the parent residue in the list.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>cutoff</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a><em>, </em><em>optional</em>) – Maximum inter-atom distance for residue to be included.
Defaults to 4.</li>
<li><strong>include_neighbours</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a><em>, </em><em>optional</em>) – If <cite>false</cite>, does not return <cite>Residue</cite> at i-1, i+1 positions
in same chain as <cite>Residue</cite>.</li>
<li><strong>inter_chain</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a><em>, </em><em>optional</em>) – If <cite>false</cite>, only includes nearby <cite>Residue</cite> in the same chain
as the <cite>Residue</cite>.</li>
<li><strong>include_ligands</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a><em>, </em><em>optional</em>) – If <cite>true</cite>, <cite>Residue</cite> classed as ligands but not identified as
solvent will be included in the environment.</li>
<li><strong>include_solvent</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a><em>, </em><em>optional</em>) – If <cite>true</cite>, Monomers classed as categorised as solvent
will be included in the environment.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>sc_environment</strong> – List of monomers within cutoff distance of side-chain.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="isambard.ampal.protein.Residue.unique_id">
<code class="descname">unique_id</code><a class="headerlink" href="#isambard.ampal.protein.Residue.unique_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a tuple that uniquely identifies a <cite>Monomer</cite> in an <cite>Assembly</cite>.</p>
<p class="rubric">Notes</p>
<p>The unique_id will uniquely identify each monomer within a polymer.
If each polymer in an assembly has a distinct id, it will uniquely
identify each monomer within the assembly.</p>
<p>The hetero-flag is defined as in Biopython as a string that is
either a single whitespace in the case of a non-hetero atom,
or ‘<a href="#id6"><span class="problematic" id="id7">H_</span></a>’ plus the name of the hetero-residue (e.g. ‘H_GLC’ in
the case of a glucose molecule), or ‘W’ in the case of a water
molecule.</p>
<p>For more information, see the Biopython documentation or this
Biopython wiki page:
<a class="reference external" href="http://biopython.org/wiki/The_Biopython_Structural_Bioinformatics_FAQ">http://biopython.org/wiki/The_Biopython_Structural_Bioinformatics_FAQ</a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><strong>unique_id</strong> – unique_id[0] is the polymer_id unique_id[1] is a triple
of the hetero-flag, the monomer id (residue number) and the
insertion code.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/2/library/functions.html#tuple" title="(in Python v2.7)">tuple</a></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="isambard.ampal.protein.align">
<code class="descname">align</code><span class="sig-paren">(</span><em>target</em>, <em>mobile</em>, <em>target_i=0</em>, <em>mobile_i=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#align"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.align" title="Permalink to this definition">¶</a></dt>
<dd><p>Aligns one Polypeptide (mobile) to another (target).</p>
<p class="rubric">Notes</p>
<p>This function directly modifies atoms of the mobile Polypeptide!
It does not return a new object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>target</strong> (<a class="reference internal" href="#isambard.ampal.protein.Polypeptide" title="isambard.ampal.protein.Polypeptide"><em>Polypeptide</em></a>) – Polypeptide to be aligned to.</li>
<li><strong>mobile</strong> (<a class="reference internal" href="#isambard.ampal.protein.Polypeptide" title="isambard.ampal.protein.Polypeptide"><em>Polypeptide</em></a>) – Polypeptide to be moved during alignment.</li>
<li><strong>target_i</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a><em>, </em><em>optional</em>) – Index of <cite>Residue</cite> in target to align to.</li>
<li><strong>mobile_i</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a><em>, </em><em>optional</em>) – Index of <cite>Residue</cite> in mobile to be aligned.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="isambard.ampal.protein.find_ss_regions_polymer">
<code class="descname">find_ss_regions_polymer</code><span class="sig-paren">(</span><em>polymer</em>, <em>ss</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#find_ss_regions_polymer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.find_ss_regions_polymer" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an <cite>Assembly</cite> of regions tagged as secondary structure.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>polymer</strong> (<a class="reference internal" href="#isambard.ampal.protein.Polypeptide" title="isambard.ampal.protein.Polypeptide"><em>Polypeptide</em></a>) – <cite>Polymer</cite> object to be searched secondary structure regions.</li>
<li><strong>ss</strong> (<em>list</em>) – List of secondary structure tags to be separate i.e. [‘H’]
would return helices, [‘H’, ‘E’] would return helices
and strands.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>fragments</strong> – <cite>Assembly</cite> containing a <cite>Polymer</cite> for each region of specified
secondary structure.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="isambard.ampal.assembly.html#isambard.ampal.assembly.Assembly" title="isambard.ampal.assembly.Assembly">Assembly</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="isambard.ampal.protein.flat_list_to_dummy_chain">
<code class="descname">flat_list_to_dummy_chain</code><span class="sig-paren">(</span><em>atom_list</em>, <em>atom_group_s=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#flat_list_to_dummy_chain"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.flat_list_to_dummy_chain" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts flat list of coordinates into dummy C-alpha carbons</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>atom_list</strong> (<em>[</em><a class="reference internal" href="isambard.ampal.base_ampal.html#isambard.ampal.base_ampal.Atom" title="isambard.ampal.base_ampal.Atom"><em>Atom</em></a><em>]</em>) – Flat list of co-ordinates.</li>
<li><strong>atom_group_s</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a><em>, </em><em>optional</em>) – Size of atom groups.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>polymer</strong> – <cite>Polymer</cite> object containing atom coord converted <cite>Monomers</cite>
with ‘DUM’ atom name.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#isambard.ampal.protein.Polypeptide" title="isambard.ampal.protein.Polypeptide">Polypeptide</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="isambard.ampal.protein.flat_list_to_polymer">
<code class="descname">flat_list_to_polymer</code><span class="sig-paren">(</span><em>atom_list</em>, <em>atom_group_s=4</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/isambard/ampal/protein.html#flat_list_to_polymer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#isambard.ampal.protein.flat_list_to_polymer" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a flat list of atomic coordinates and converts it to a <cite>Polymer</cite>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>atom_list</strong> (<em>[</em><a class="reference internal" href="isambard.ampal.base_ampal.html#isambard.ampal.base_ampal.Atom" title="isambard.ampal.base_ampal.Atom"><em>Atom</em></a><em>]</em>) – Flat list of coordinates.</li>
<li><strong>atom_group_s</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a><em>, </em><em>optional</em>) – Size of atom groups.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>polymer</strong> – <cite>Polymer</cite> object containing atom coords converted <cite>Monomers</cite>.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><a class="reference internal" href="#isambard.ampal.protein.Polypeptide" title="isambard.ampal.protein.Polypeptide">Polypeptide</a></p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> – Raised if <cite>atom_group_s</cite> != 4 or 5</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="isambard.ampal.pseudo_atoms.html" class="btn btn-neutral float-right" title="isambard.ampal.pseudo_atoms module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="isambard.ampal.pdb_parser.html" class="btn btn-neutral" title="isambard.ampal.pdb_parser module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Woolfson Group.
      Last updated on August 17, 2017.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.4.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>